// @entelligentsia/forgecli/dist/extensions/forgecli/config-tui/state/init.js
// Config TUI state constructor — pure function, no I/O.
// Split from state.ts (Phase 1).
export function initialState(opts) {
    const buffer = {
        global: opts.global ? cloneJSON(opts.global) : {},
        project: opts.project ? cloneJSON(opts.project) : {},
    };
    const firstView = { kind: "tier-menu", cursor: 0 };
    // Default scope: project when in a project, global otherwise.
    const scope = opts.pipelineCatalogue !== null ? "project" : "global";
    return {
        buffer,
        view: [firstView],
        cwd: opts.cwd,
        personaCatalogue: opts.personaCatalogue,
        pipelineCatalogue: opts.pipelineCatalogue,
        availableModels: opts.availableModels,
        authenticatedProviders: opts.authenticatedProviders,
        dirty: false,
        scope,
        authError: opts.authError ?? null,
        configErrors: opts.configErrors && opts.configErrors.length > 0 ? opts.configErrors : null,
        confirmQuit: false,
        shouldExit: false,
        lastSaved: null,
    };
}
function cloneJSON(value) {
    return JSON.parse(JSON.stringify(value));
}
