// @entelligentsia/forgecli/dist/extensions/forgecli/config-tui/screens/overrides-list.js
// Overrides-list-pipelines screen — list of pipelines with override counts.
// Phase 3: full theming, width safety.
import { Key, matchesKey } from "@earendil-works/pi-tui";
import { getActiveView, listPipelineOverrideSummaries } from "../state/selectors.js";
import { accentBold, dirtyMarker, muted, padRight } from "../theme.js";
import { rule, safeLines, windowList } from "./shared.js";
export class OverridesListPipelinesScreen {
    render(state, width, theme) {
        const view = getActiveView(state);
        if (view.kind !== "overrides-list-pipelines") {
            return ["(renderOverridesListPipelines called with wrong active view)"];
        }
        const summaries = listPipelineOverrideSummaries(state);
        const lines = [];
        lines.push(accentBold("forge config › per-step overrides", theme));
        lines.push(rule(width, theme));
        lines.push(muted("  Step overrides change the model for one specific step", theme));
        lines.push(muted("  of one pipeline. Most users don't need this.", theme));
        lines.push("");
        if (summaries.length === 0) {
            lines.push(muted("  (no pipeline catalogue available)", theme));
            lines.push(muted("  esc back", theme));
            return safeLines(lines, width);
        }
        lines.push(muted("  Pipelines:", theme));
        const win = windowList(summaries, view.cursor);
        if (win.aboveCount > 0)
            lines.push(muted(`    ↑ ${win.aboveCount} more above`, theme));
        win.visible.forEach((s, i) => {
            const absoluteIdx = win.start + i;
            const cur = absoluteIdx === view.cursor ? theme.fg("accent", "▸") : " ";
            const status = s.overrideCount === 0
                ? muted("none", theme)
                : accentBold(`${s.overrideCount} override${s.overrideCount === 1 ? "" : "s"}`, theme);
            lines.push(`  ${cur} ${padRight(s.pipeline, 18)} ${status}`);
        });
        if (win.belowCount > 0)
            lines.push(muted(`    ↓ ${win.belowCount} more below`, theme));
        lines.push("");
        lines.push(muted("  enter inspect highlighted pipeline   esc back", theme));
        if (state.dirty)
            lines.push(`  ${dirtyMarker(theme)}`);
        return safeLines(lines, width);
    }
    handleInput(data, state) {
        const view = getActiveView(state);
        if (view.kind !== "overrides-list-pipelines")
            return { kind: "no-op" };
        const summaries = listPipelineOverrideSummaries(state);
        if (matchesKey(data, Key.up) || matchesKey(data, "k")) {
            return { kind: "dispatch", action: { kind: "cursor-move", delta: -1 } };
        }
        if (matchesKey(data, Key.down) || matchesKey(data, "j")) {
            if (view.cursor < summaries.length - 1)
                return { kind: "dispatch", action: { kind: "cursor-move", delta: 1 } };
            return { kind: "consumed" };
        }
        if (matchesKey(data, Key.enter)) {
            const target = summaries[view.cursor];
            if (target) {
                return {
                    kind: "dispatch",
                    action: {
                        kind: "push-view",
                        view: { kind: "overrides-list-phases", pipeline: target.pipeline, cursor: 0 },
                    },
                };
            }
            return { kind: "consumed" };
        }
        return { kind: "no-op" };
    }
}
