// @entelligentsia/forge-compress/src/forge/markdown.ts
import { countTokens, truncateToTokenBudget } from "../tokens.js";

export interface CompressMarkdownOptions {
  maxTokens?: number;
  mode?: "map" | "headings" | "truncate";
}

export function compressMarkdown(
  input: string,
  opts?: CompressMarkdownOptions,
): string {
  if (!input.trim()) return "";

  const mode = opts?.mode ?? "map";
  const maxTokens = opts?.maxTokens ?? Infinity;

  switch (mode) {
    case "headings":
      return compressHeadings(input, maxTokens);
    case "truncate":
      return truncateToTokenBudget(input, maxTokens);
    case "map":
    default:
      return compressMap(input, maxTokens);
  }
}

function compressMap(input: string, maxTokens: number): string {
  const lines = input.split("\n");
  const picked: string[] = [];
  let inFrontmatter = false;
  let frontmatterDone = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const trimmed = line.trim();

    if (i === 0 && trimmed === "---") {
      inFrontmatter = true;
      continue;
    }
    if (inFrontmatter) {
      if (trimmed === "---") {
        inFrontmatter = false;
        frontmatterDone = true;
      }
      continue;
    }

    if (trimmed.startsWith("#")) {
      picked.push(line);
      continue;
    }

    if (
      trimmed.startsWith("- **") ||
      trimmed.startsWith("* **") ||
      trimmed.startsWith("| ") ||
      trimmed.startsWith("```")
    ) {
      picked.push(line);
      continue;
    }

    if (
      trimmed.startsWith("- [ ]") ||
      trimmed.startsWith("- [x]") ||
      trimmed.startsWith("- [X]")
    ) {
      picked.push(line);
      continue;
    }

    if (i === 0 || (frontmatterDone && picked.length === 0)) {
      picked.push(line);
    }
  }

  if (picked.length === 0) {
    return truncateToTokenBudget(input, maxTokens);
  }

  const totalLines = lines.length;
  const keptLines = picked.length;
  const omitted = totalLines - keptLines;

  let result = picked.join("\n");
  if (omitted > 0) {
    result += `\n\n[${omitted} body lines omitted — ${totalLines} total]`;
  }

  if (countTokens(result) > maxTokens) {
    result = truncateToTokenBudget(result, maxTokens);
  }

  return result;
}

function compressHeadings(input: string, maxTokens: number): string {
  const lines = input.split("\n");
  const headings: string[] = [];
  let depth = 0;

  for (const line of lines) {
    const match = line.match(/^(#{1,6})\s+(.+)/);
    if (match) {
      const level = match[1].length;
      const indent = "  ".repeat(level - 1);
      headings.push(`${indent}${match[2]}`);
      depth = Math.max(depth, level);
    }
  }

  if (headings.length === 0) {
    return truncateToTokenBudget(input, maxTokens);
  }

  const result = `${headings.length} sections (depth ${depth}):\n${headings.join("\n")}`;
  return countTokens(result) > maxTokens
    ? truncateToTokenBudget(result, maxTokens)
    : result;
}
