// @entelligentsia/forge-compress/src/forge/query.ts
import { countTokens } from "../tokens.js";

interface QueryResultEntry {
  id: string;
  title: string;
  status: string;
  type: string;
  relationships?: Record<string, unknown>;
  fileRefs?: { json: string; md: string };
  storeRef?: string;
  indexRef?: string;
  excerpt?: string | null;
  [key: string]: unknown;
}

interface QueryEnvelope {
  query?: string;
  path?: string;
  traversalTrace?: string[];
  results?: QueryResultEntry[];
  relatedFileRefs?: string[];
  count?: number;
  totalMatched?: number;
  returned?: number;
  limit?: number | null;
  sort?: string | null;
  config?: Record<string, unknown>;
  meta?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface CompressQueryOptions {
  keepExcerpts?: boolean;
  maxExcerptLines?: number;
  maxResults?: number;
  maxTokens?: number;
}

export function compressStoreQuery(
  input: string,
  opts?: CompressQueryOptions,
): string {
  let envelope: QueryEnvelope;
  try {
    envelope = JSON.parse(input);
  } catch {
    return input;
  }

  if (!envelope.results || !Array.isArray(envelope.results)) return input;

  const keepExcerpts = opts?.keepExcerpts ?? false;
  const maxExcerptLines = opts?.maxExcerptLines ?? 5;
  const maxResults = opts?.maxResults ?? envelope.results.length;

  const results = envelope.results.slice(0, maxResults).map((r) => {
    const entry: Record<string, unknown> = {
      id: r.id,
      title: r.title,
      status: r.status,
      type: r.type,
    };

    if (r.relationships && Object.keys(r.relationships).length > 0) {
      const rels: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(r.relationships)) {
        if (v !== null && v !== undefined) rels[k] = v;
      }
      if (Object.keys(rels).length > 0) entry.relationships = rels;
    }

    if (keepExcerpts && r.excerpt) {
      const lines = r.excerpt.split("\n");
      entry.excerpt =
        lines.length > maxExcerptLines
          ? lines.slice(0, maxExcerptLines).join("\n") +
            `\n... (${lines.length - maxExcerptLines} more lines)`
          : r.excerpt;
    }

    return entry;
  });

  const compact: Record<string, unknown> = { results };

  if (envelope.query) compact.query = envelope.query;
  if (envelope.path) compact.path = envelope.path;
  if (typeof envelope.count === "number") compact.count = envelope.count;
  if (typeof envelope.totalMatched === "number")
    compact.totalMatched = envelope.totalMatched;
  if (typeof envelope.returned === "number")
    compact.returned = envelope.returned;

  if (envelope.results.length > maxResults) {
    compact.truncated = envelope.results.length - maxResults;
  }

  let output = JSON.stringify(compact, null, 2);

  if (opts?.maxTokens && countTokens(output) > opts.maxTokens) {
    output = JSON.stringify(compact);
  }

  return output;
}
