// @entelligentsia/forgecli/dist/extensions/forgecli/dashboard/component.js
// dashboard/component.ts — Two-panel orchestrator tree dashboard.
//
// Renders as a pi-tui overlay (ctx.ui.custom) with:
//   Left panel:  tree browser with depth-based indentation, status glyphs,
//                progress labels, and cursor highlighting.
//   Right panel: detail view of the selected node — status, model, metrics,
//                prompt preview, tail-buffer activity, outcome, and children.
//
// Keyboard:
//   ↑/↓     move cursor in tree
//   →       expand orchestrator / focus leaf detail
//   ←       collapse orchestrator / back to tree
//   Enter   toggle expand on orchestrator; focus detail on leaf
//   Esc     close overlay
//   x       request cancellation (with y/n confirm)
//   p       (reserved for pause)
//
// The component reads from OrchestratorTree (the model) and subscribes to
// its events for live updates. No writes flow through the component —
// the model is mutated by the orchestrator code, not the dashboard.
import { matchesKey, Key, truncateToWidth, visibleWidth } from "@earendil-works/pi-tui";
import { fmtTokenMeter } from "../viewport-renderer.js";
import { paintTailLine } from "../viewport-theme.js";
// ── Braille spinner ─────────────────────────────────────────────────────────
const SPINNER_FRAMES = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
const SPINNER_INTERVAL_MS = 100;
// ── Controller ──────────────────────────────────────────────────────────────
export class DashboardController {
    tree;
    state;
    /** Callback to request a TUI re-render after state changes. */
    onInvalidate;
    constructor(tree, initialCursorId) {
        this.tree = tree;
        // Default cursor to the first root, or empty string if tree is empty.
        const roots = tree.getRoots();
        const firstVisible = roots.length > 0 ? roots[0].id : "";
        this.state = {
            cursorId: initialCursorId ?? firstVisible,
            expanded: new Set(),
            focusPanel: "tree",
            promptExpanded: false,
            cancelTargetId: null,
            detailScroll: 0,
        };
    }
    getState() {
        return this.state;
    }
    setOnInvalidate(cb) {
        this.onInvalidate = cb;
    }
    // ── Visible node list (DFS respecting expand state) ────────────────────
    getVisibleNodes() {
        const result = [];
        const visit = (id) => {
            const node = this.tree.getNode(id);
            if (!node)
                return;
            result.push(id);
            if (node.kind === "orchestrator" && this.state.expanded.has(id)) {
                for (const childId of node.children) {
                    visit(childId);
                }
            }
        };
        for (const root of this.tree.getRoots()) {
            visit(root.id);
        }
        return result;
    }
    // ── Input handling ──────────────────────────────────────────────────────
    handleInput(data) {
        // Cancel confirmation mode takes priority.
        if (this.state.cancelTargetId) {
            this.handleCancelConfirm(data);
            this.onInvalidate?.();
            return;
        }
        switch (this.state.focusPanel) {
            case "tree":
                this.handleTreeInput(data);
                break;
            case "detail":
                this.handleDetailInput(data);
                break;
        }
        this.onInvalidate?.();
    }
    handleTreeInput(data) {
        if (matchesKey(data, Key.up)) {
            this.moveCursor(-1);
        }
        else if (matchesKey(data, Key.down)) {
            this.moveCursor(1);
        }
        else if (matchesKey(data, Key.enter) || matchesKey(data, Key.right)) {
            this.activateCursor();
        }
        else if (matchesKey(data, Key.left)) {
            this.collapseCursor();
        }
        else if (data === "x") {
            this.startCancel();
        }
        else if (matchesKey(data, Key.escape)) {
            // Esc in tree = request close. The done callback handles it.
            // Mark a sentinel so the mount code can close.
            this.state.cursorId = "__close__";
        }
        // Consume all keys while overlay is active (no passthrough).
    }
    handleDetailInput(data) {
        if (matchesKey(data, Key.escape) || matchesKey(data, Key.left)) {
            this.state.focusPanel = "tree";
            this.state.detailScroll = 0;
        }
        else if (matchesKey(data, Key.down) || data === "j" || data === "J") {
            this.state.detailScroll++;
        }
        else if (matchesKey(data, Key.up) || data === "k" || data === "K") {
            this.state.detailScroll = Math.max(0, this.state.detailScroll - 1);
        }
        else if (matchesKey(data, Key.enter)) {
            // Toggle prompt expansion when in detail view
            this.state.promptExpanded = !this.state.promptExpanded;
        }
        else if (data === "x") {
            this.startCancel();
        }
    }
    handleCancelConfirm(data) {
        if (data === "y" || matchesKey(data, Key.enter)) {
            if (this.state.cancelTargetId) {
                this.tree.requestCancel(this.state.cancelTargetId);
            }
            this.state.cancelTargetId = null;
        }
        else if (data === "n" || matchesKey(data, Key.escape)) {
            this.state.cancelTargetId = null;
        }
        // All other keys consumed silently in cancel mode.
    }
    moveCursor(delta) {
        const visible = this.getVisibleNodes();
        if (visible.length === 0)
            return;
        const idx = visible.indexOf(this.state.cursorId);
        const next = Math.max(0, Math.min(visible.length - 1, idx + delta));
        this.state.cursorId = visible[next];
        // Auto-expand ancestors so the cursor is always visible.
        this.ensureAncestorsExpanded(this.state.cursorId);
    }
    activateCursor() {
        const node = this.tree.getNode(this.state.cursorId);
        if (!node)
            return;
        if (node.kind === "orchestrator") {
            this.toggleExpand(node.id);
        }
        else {
            this.state.focusPanel = "detail";
            this.state.detailScroll = 0;
        }
    }
    toggleExpand(id) {
        if (this.state.expanded.has(id)) {
            this.state.expanded.delete(id);
        }
        else {
            this.state.expanded.add(id);
        }
    }
    collapseCursor() {
        const node = this.tree.getNode(this.state.cursorId);
        if (!node)
            return;
        if (node.kind === "orchestrator" && this.state.expanded.has(node.id)) {
            this.state.expanded.delete(node.id);
        }
        else if (node.parentId) {
            // Move cursor to parent.
            this.state.cursorId = node.parentId;
        }
    }
    startCancel() {
        const node = this.tree.getNode(this.state.cursorId);
        if (!node)
            return;
        // Can cancel running nodes, or the root orchestrator session.
        if (node.status === "running" || node.status === "cancelling") {
            this.state.cancelTargetId = node.id;
        }
    }
    ensureAncestorsExpanded(id) {
        for (const ancestor of this.tree.getAncestors(id)) {
            if (ancestor.kind === "orchestrator") {
                this.state.expanded.add(ancestor.id);
            }
        }
    }
    // ── Auto-expand: when a new running node appears, auto-expand its ────
    // parent so it's immediately visible in the tree.
    autoExpandNewNode(id) {
        const node = this.tree.getNode(id);
        if (!node)
            return;
        if (node.kind === "orchestrator") {
            this.state.expanded.add(id);
        }
        if (node.parentId) {
            const parent = this.tree.getNode(node.parentId);
            if (parent?.kind === "orchestrator") {
                this.state.expanded.add(parent.id);
            }
        }
        // Default cursor to the newest running node if no cursor set.
        if (!this.state.cursorId || this.state.cursorId === "") {
            this.state.cursorId = id;
        }
    }
    isCloseRequested() {
        return this.state.cursorId === "__close__";
    }
}
// ── View ────────────────────────────────────────────────────────────────────
export class DashboardComponent {
    tree;
    controller;
    theme;
    tui;
    done;
    spinnerTimer;
    _rerender;
    constructor(tree, tui, theme, done) {
        this.tree = tree;
        this.tui = tui;
        this.theme = theme;
        this.done = done;
        this.controller = new DashboardController(tree);
        const rerender = () => this.tui.requestRender();
        this.tree.on("change", rerender);
        this.tree.on("tail", rerender);
        this.tree.on("preview", rerender);
        this.controller.setOnInvalidate(rerender);
        // Spinner timer: re-render at 100ms intervals while any node is running.
        this.ensureSpinnerTimer();
        this._rerender = rerender;
    }
    ensureSpinnerTimer() {
        if (this.spinnerTimer)
            return;
        this.spinnerTimer = setInterval(() => {
            const anyRunning = this.tree.getRoots().some((r) => r.status === "running" || r.status === "cancelling");
            if (!anyRunning) {
                // One last render to settle spinner into final frame.
                this.tui.requestRender();
                if (this.spinnerTimer)
                    clearInterval(this.spinnerTimer);
                this.spinnerTimer = undefined;
                return;
            }
            this.tui.requestRender();
        }, SPINNER_INTERVAL_MS);
    }
    // ── Component interface ──────────────────────────────────────────────────
    render(width) {
        // Check if close was requested.
        if (this.controller.isCloseRequested()) {
            this.dispose();
            this.done(null);
            return [];
        }
        // Layout: left panel 30%, right panel 70%, borders.
        const leftWidth = Math.max(30, Math.floor(width * 0.30));
        const separatorWidth = 1;
        const rightWidth = Math.max(20, width - leftWidth - separatorWidth - 3);
        const contentWidth = width - 2; // minus left and right border chars
        const visible = this.controller.getVisibleNodes();
        const state = this.controller.getState();
        const selectedNode = this.tree.getNode(state.cursorId);
        // ── Cancel confirmation overlay ────────────────────────────────
        if (state.cancelTargetId) {
            return this.renderCancelConfirm(width, state.cancelTargetId);
        }
        // ── Left panel: tree browser ───────────────────────────────────
        const leftLines = this.renderTreePanel(visible, state, leftWidth);
        // ── Right panel: detail ─────────────────────────────────────────
        const rightLines = selectedNode
            ? this.renderDetailPanel(selectedNode, rightWidth)
            : [this.theme.fg("dim", "Select a node in the tree")];
        // ── Compose ─────────────────────────────────────────────────────
        const termHeight = this.tui.terminal.rows;
        // Deduct 3 lines for top border, bottom border, and hints footer
        const contentHeight = Math.max(8, termHeight - 3);
        // Compute detail panel scroll with auto-scrolling fallback for tree focus
        let activeScroll = state.detailScroll;
        if (state.focusPanel === "tree") {
            // Auto-scroll to the bottom so live dispatches are visible immediately
            activeScroll = Math.max(0, rightLines.length - contentHeight);
        }
        else {
            // Clamp user manual scroll within valid bounds
            const maxScroll = Math.max(0, rightLines.length - contentHeight);
            if (activeScroll > maxScroll) {
                activeScroll = maxScroll;
                state.detailScroll = maxScroll; // sync controller state
            }
        }
        const paddedLeftLines = [];
        for (let i = 0; i < contentHeight; i++) {
            paddedLeftLines.push(leftLines[i] ?? "");
        }
        const paddedRightLines = [];
        for (let i = 0; i < contentHeight; i++) {
            paddedRightLines.push(rightLines[i + activeScroll] ?? "");
        }
        const lines = [];
        const border = (s) => this.theme.fg("border", s);
        const accent = (s) => this.theme.fg("accent", s);
        // ── Header ──────────────────────────────────────────────────────
        const headerText = ` Orchestrator Dashboard `;
        const headerPad = contentWidth - visibleWidth(border("")) - visibleWidth(accent(headerText));
        lines.push(border("╭") +
            accent(headerText) +
            border("─".repeat(Math.max(0, headerPad))) +
            border("╮"));
        for (let i = 0; i < contentHeight; i++) {
            const left = truncateToWidth(paddedLeftLines[i], leftWidth);
            const right = truncateToWidth(paddedRightLines[i], rightWidth);
            const lPad = leftWidth - visibleWidth(left);
            const rPad = rightWidth - visibleWidth(right);
            lines.push(border("│") +
                left +
                " ".repeat(Math.max(0, lPad)) +
                border("│") +
                " " +
                right +
                " ".repeat(Math.max(0, rPad)) +
                border("│"));
        }
        // ── Footer with key hints ───────────────────────────────────────
        const dim = (s) => this.theme.fg("dim", s);
        const hints = " ↑↓ select · → expand · ← collapse · ⏎ focus · x stop · esc close";
        lines.push(border("╰") + border("─".repeat(contentWidth)) + border("╯"));
        lines.push(dim(truncateToWidth(hints, width)));
        return lines;
    }
    handleInput(data) {
        this.controller.handleInput(data);
        this.tui.requestRender();
    }
    invalidate() {
        // No cached render state — pure function of tree + state.
    }
    dispose() {
        if (this.spinnerTimer) {
            clearInterval(this.spinnerTimer);
            this.spinnerTimer = undefined;
        }
        if (this._rerender) {
            this.tree.off("change", this._rerender);
            this.tree.off("tail", this._rerender);
            this.tree.off("preview", this._rerender);
        }
    }
    // ── Tree panel renderer ─────────────────────────────────────────────────
    renderTreePanel(visibleIds, state, width) {
        if (visibleIds.length === 0) {
            return [this.theme.fg("dim", "No session running")];
        }
        const lines = [];
        // Header
        lines.push(this.theme.bold(this.theme.fg("accent", " Phases")));
        for (const id of visibleIds) {
            const node = this.tree.getNode(id);
            if (!node)
                continue;
            const isCursor = id === state.cursorId;
            const depth = this.tree.getDepth(id);
            const indent = " ".repeat(depth * 2);
            // Status glyph
            const glyph = this.nodeGlyph(node);
            // Progress label for orchestrators
            let label = node.label;
            if (node.kind === "orchestrator") {
                const prog = this.tree.getSubtreeProgress(id);
                label += ` ${prog.completed}/${prog.total}`;
            }
            // Expand indicator
            const expandIndicator = node.kind === "orchestrator"
                ? state.expanded.has(id)
                    ? " "
                    : "▸"
                : " ";
            // Combine
            const prefix = `${indent}${isCursor ? "❯" : " "} ${glyph} ${expandIndicator} `;
            const styled = isCursor
                ? this.theme.bold(this.theme.fg("accent", `${prefix}${label}`))
                : `${prefix}${label}`;
            lines.push(truncateToWidth(styled, width));
        }
        return lines;
    }
    nodeGlyph(node) {
        switch (node.status) {
            case "completed":
                return this.theme.fg("success", "✔");
            case "running":
                return this.theme.fg("accent", this.spinnerFrame());
            case "cancelling":
                return this.theme.fg("warning", "⏳");
            case "cancelled":
                return this.theme.fg("muted", "⊘");
            case "failed":
                return this.theme.fg("error", "✗");
            case "escalated":
                return this.theme.fg("error", "▲");
            case "pending":
            default:
                return this.theme.fg("dim", "○");
        }
    }
    spinnerFrame() {
        const idx = Math.floor(Date.now() / SPINNER_INTERVAL_MS) % SPINNER_FRAMES.length;
        return SPINNER_FRAMES[idx];
    }
    // ── Detail panel renderer ────────────────────────────────────────────────
    renderDetailPanel(node, width) {
        const lines = [];
        const dim = (s) => this.theme.fg("dim", s);
        const accent = (s) => this.theme.fg("accent", s);
        const bold = (s) => this.theme.bold(s);
        const success = (s) => this.theme.fg("success", s);
        const warn = (s) => this.theme.fg("warning", s);
        // ── Header: label + status ───────────────────────────────────────
        const statusLabel = this.statusLabel(node.status);
        const modelPart = node.model ? ` · ${node.provider ?? ""} ${node.model}` : "";
        lines.push(`${this.nodeGlyph(node)} ${bold(statusLabel)}${dim(modelPart)}`);
        // ── Metrics line ────────────────────────────────────────────────
        const metrics = this.formatMetrics(node);
        if (metrics)
            lines.push(dim(metrics));
        lines.push("");
        // ── Orchestrator node: list children ─────────────────────────────
        if (node.kind === "orchestrator" && node.children.length > 0) {
            const children = this.tree.getChildren(node.id);
            lines.push(dim(bold(`Agents · ${children.length}`)));
            for (const child of children) {
                const cglyph = this.nodeGlyph(child);
                const cmodel = child.model ? ` ${child.provider ?? ""} ${child.model}` : "";
                const cmetrics = this.formatMetrics(child);
                const cmetricsPart = cmetrics ? ` ${cmetrics}` : "";
                lines.push(` ${cglyph} ${child.label}${dim(cmodel)}${dim(cmetricsPart)}`);
            }
            lines.push("");
        }
        // ── Prompt preview (expandable) ─────────────────────────────────
        if (node.promptPreview) {
            const expandIcon = this.controller.getState().promptExpanded ? "▼" : "▶";
            const lineCount = node.promptPreview.split("\n").length;
            lines.push(dim(`${expandIcon} Prompt · ${lineCount} lines · ⏎ expand`));
            if (this.controller.getState().promptExpanded) {
                for (const pline of node.promptPreview.split("\n").slice(0, 20)) {
                    lines.push(dim(`  ${truncateToWidth(pline, width - 4)}`));
                }
                if (lineCount > 20) {
                    lines.push(dim(`  … ${lineCount - 20} more lines`));
                }
            }
            lines.push("");
        }
        // ── Activity: running full log of all turns ──────────────────────
        if (node.kind === "leaf" && node.tailBuffer.length > 0) {
            const total = node.tailBuffer.length;
            lines.push(dim(`Activity · ${total} log line${total === 1 ? "" : "s"}`));
            for (const tline of node.tailBuffer) {
                const painted = paintTailLine(tline, this.theme);
                lines.push(truncateToWidth(painted, width));
            }
            lines.push("");
        }
        // ── Outcome ─────────────────────────────────────────────────────
        if (node.outcomePreview) {
            lines.push(dim("Outcome"));
            for (const oline of node.outcomePreview.split("\n").slice(0, 8)) {
                lines.push(`  ${truncateToWidth(oline, width - 4)}`);
            }
            const lineCount = node.outcomePreview.split("\n").length;
            if (lineCount > 8) {
                lines.push(dim(`  … ${lineCount - 8} more lines`));
            }
        }
        return lines;
    }
    renderCancelConfirm(width, targetId) {
        const node = this.tree.getNode(targetId);
        const label = node?.label ?? targetId;
        const dim = (s) => this.theme.fg("dim", s);
        const warn = (s) => this.theme.fg("warning", s);
        const bold = (s) => this.theme.bold(s);
        const border = (s) => this.theme.fg("border", s);
        const prompt = warn(`⚠ Cancel ${bold(label)}?`);
        const actions = dim("y confirm · n/esc dismiss");
        const padW = Math.max(0, width - visibleWidth(prompt) - visibleWidth(actions) - 4);
        const lines = [];
        lines.push(border("╭" + "─".repeat(width - 2) + "╮"));
        lines.push(border("│") + " ".repeat(width - 2) + border("│"));
        lines.push(border("│") +
            " " +
            prompt +
            " ".repeat(Math.max(0, padW)) +
            " " +
            actions +
            " ".repeat(Math.max(0, width - 4 - visibleWidth(prompt) - visibleWidth(actions) - padW)) +
            border("│"));
        lines.push(border("│") + " ".repeat(width - 2) + border("│"));
        lines.push(border("╰" + "─".repeat(width - 2) + "╯"));
        return lines;
    }
    statusLabel(status) {
        switch (status) {
            case "completed":
                return "Completed";
            case "running":
                return "Running";
            case "cancelling":
                return "Cancelling…";
            case "cancelled":
                return "Cancelled";
            case "failed":
                return "Failed";
            case "escalated":
                return "Escalated";
            case "pending":
                return "Pending";
        }
    }
    formatMetrics(node) {
        const parts = [];
        if (node.usage.input || node.usage.output || node.usage.cacheRead) {
            parts.push(fmtTokenMeter(node.usage));
        }
        if (node.metrics.toolCount) {
            parts.push(`${node.metrics.toolCount} tool${node.metrics.toolCount === 1 ? "" : "s"}`);
        }
        if (node.metrics.errCount) {
            parts.push(`${node.metrics.errCount} err`);
        }
        if (node.startedAt) {
            const end = node.endedAt ?? Date.now();
            const secs = Math.floor((end - node.startedAt) / 1000);
            parts.push(secs < 60 ? `${secs}s` : `${Math.floor(secs / 60)}m${String(secs % 60).padStart(2, "0")}s`);
        }
        return parts.join(" · ");
    }
}
