// @entelligentsia/forgecli/dist/extensions/forgecli/config-tui/plugin-config-reader.js
// Read-only loader for the Forge plugin's `.forge/config.json`.
//
// Plan 16 Slice 4a: the config TUI needs to render a read-only summary of the
// installed Forge plugin (version, paths, installed skills) alongside the
// forge-cli routing config it edits. This module never mutates `.forge/`.
//
// The legacy `mode` field (fast/full) is retired (see memory
// project_fast_full_mode_retired). When present, we surface it as
// modeLegacy=true so the TUI can render it as "(legacy)" without offering to
// change it.
import * as fs from "node:fs";
import * as path from "node:path";
export function readPluginConfig(cwd) {
    const configPath = path.join(cwd, ".forge", "config.json");
    let raw;
    try {
        raw = fs.readFileSync(configPath, "utf-8");
    }
    catch {
        return null;
    }
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch {
        return null;
    }
    if (typeof parsed !== "object" || parsed === null)
        return null;
    const obj = parsed;
    const project = (obj.project ?? null);
    const paths = (obj.paths ?? {});
    const installedSkills = Array.isArray(obj.installedSkills)
        ? obj.installedSkills.filter((s) => typeof s === "string")
        : [];
    const mode = typeof obj.mode === "string" ? obj.mode : null;
    return {
        version: typeof obj.version === "string" ? obj.version : null,
        projectName: typeof project?.name === "string" ? project.name : null,
        projectPrefix: typeof project?.prefix === "string" ? project.prefix : null,
        paths: {
            engineering: typeof paths.engineering === "string" ? paths.engineering : undefined,
            store: typeof paths.store === "string" ? paths.store : undefined,
            workflows: typeof paths.workflows === "string" ? paths.workflows : undefined,
            commands: typeof paths.commands === "string" ? paths.commands : undefined,
            templates: typeof paths.templates === "string" ? paths.templates : undefined,
            customCommands: typeof paths.customCommands === "string" ? paths.customCommands : undefined,
            forgeRoot: typeof paths.forgeRoot === "string" ? paths.forgeRoot : undefined,
            forgeRef: typeof paths.forgeRef === "string" ? paths.forgeRef : undefined,
        },
        installedSkills,
        forgeRef: typeof paths.forgeRef === "string" ? paths.forgeRef : null,
        mode,
        modeLegacy: mode !== null,
    };
}
