// @entelligentsia/forgecli/dist/extensions/forgecli/config-tui/theme.js
// Config-TUI theme helpers — wraps pi Theme for consistent styling.
//
// Every render function receives a Theme instance and uses these helpers
// (or calls theme.fg/bg/bg directly) for all visible text. No raw string
// styling should appear in screen renderers — route everything through
// here or through theme methods so theme changes propagate globally.
import { truncateToWidth, visibleWidth } from "@earendil-works/pi-tui";
// ── Structural helpers ──────────────────────────────────────────────────────
/** Horizontal rule spanning the full width, themed as a border. */
export function rule(width, theme) {
    return theme.fg("border", "─".repeat(Math.max(1, width)));
}
/** Breadcrumb: "forge config › personas › pick which". */
export function breadcrumb(parts, width, theme) {
    const raw = parts.join(" › ");
    const truncated = truncateToWidth(raw, width, "");
    return theme.fg("accent", theme.bold(truncated));
}
// ── Inline style helpers ────────────────────────────────────────────────────
export function cursor(isSelected, theme) {
    return isSelected ? theme.fg("accent", "▸") : " ";
}
export function authBadge(ok, theme) {
    return ok ? theme.fg("success", "✓") : theme.fg("error", "✗");
}
export function warning(text, theme) {
    return theme.fg("warning", text);
}
export function muted(text, theme) {
    return theme.fg("muted", text);
}
export function dim(text, theme) {
    return theme.fg("dim", text);
}
export function accent(text, theme) {
    return theme.fg("accent", text);
}
export function accentBold(text, theme) {
    return theme.fg("accent", theme.bold(text));
}
export function error(text, theme) {
    return theme.fg("error", text);
}
export function success(text, theme) {
    return theme.fg("success", text);
}
/** Mark an unsaved-change indicator — themed as warning. */
export function dirtyMarker(theme) {
    return theme.fg("warning", "* unsaved");
}
/** Tier badge — displays Heavy / Standard / Light in accent color. */
export function tierBadge(tier, theme) {
    const labels = { heavy: "Heavy", standard: "Standard", light: "Light" };
    return theme.fg("accent", labels[tier] ?? tier);
}
// ── Padding (ANSI-aware) ────────────────────────────────────────────────────
/**
 * Right-pad a string that may contain ANSI escape codes to a given visible width.
 * Uses visibleWidth so ANSI codes don't inflate the padding.
 * Does NOT truncate — allows over-wide columns. Width-safety truncation
 * is applied at the line level by truncateLines() / safeLines().
 */
export function padRight(text, width) {
    const vis = visibleWidth(text);
    if (vis >= width)
        return text;
    return text + " ".repeat(width - vis);
}
/**
 * Truncate a potentially-ANSI string to a visible width, padding with spaces
 * if shorter. This replaces the old padRight for column-aligned output and
 * is safe for themed text.
 */
export function padOrTruncate(text, width) {
    const vis = visibleWidth(text);
    if (vis >= width)
        return truncateToWidth(text, width, "");
    return text + " ".repeat(width - vis);
}
// ── Width safety ─────────────────────────────────────────────────────────────
/**
 * Truncate every line to the given visible width. Applied as a final safety
 * guard in every screen's render method so no line overflows the terminal.
 */
export function truncateLines(lines, width) {
    return lines.map((line) => truncateToWidth(line, width, ""));
}
