// @entelligentsia/forgecli/dist/extensions/forgecli/wf-engine/state-store.js
import * as fs from "node:fs";
import * as path from "node:path";
export class StateStore {
    workingDir;
    constructor(workingDir) {
        this.workingDir = workingDir;
    }
    initialState(initial) {
        fs.mkdirSync(this.workingDir, { recursive: true });
        fs.mkdirSync(path.join(this.workingDir, "nodes"), { recursive: true });
        fs.mkdirSync(path.join(this.workingDir, "artifacts"), { recursive: true });
        fs.writeFileSync(this.statePath(), JSON.stringify(initial, null, 2));
        fs.writeFileSync(this.eventLogPath(), ""); // create empty file
    }
    readState() {
        return JSON.parse(fs.readFileSync(this.statePath(), "utf8"));
    }
    writeState(s) {
        // Atomic write: tmp + rename
        const tmp = `${this.statePath()}.tmp`;
        fs.writeFileSync(tmp, JSON.stringify(s, null, 2));
        fs.renameSync(tmp, this.statePath());
    }
    appendEvents(events) {
        const lines = events.map((e) => JSON.stringify(e)).join("\n") + "\n";
        fs.appendFileSync(this.eventLogPath(), lines);
    }
    writeArtifact(relPath, content) {
        const abs = path.join(this.workingDir, relPath);
        fs.mkdirSync(path.dirname(abs), { recursive: true });
        fs.writeFileSync(abs, content);
    }
    writeNodeArchive(nodeExecId, files) {
        const dir = path.join(this.workingDir, "nodes", nodeExecId);
        fs.mkdirSync(dir, { recursive: true });
        for (const [name, content] of Object.entries(files)) {
            fs.writeFileSync(path.join(dir, name), content);
        }
    }
    statePath() {
        return path.join(this.workingDir, "state.json");
    }
    eventLogPath() {
        return path.join(this.workingDir, "events.log.jsonl");
    }
}
