// @entelligentsia/forgecli/dist/extensions/forgecli/subagent/agents.js
/**
 * Vendored from pi-mono — DO NOT EDIT BEHAVIOUR HERE.
 * Source:  packages/coding-agent/examples/extensions/subagent/agents.ts
 * Repo:    https://github.com/badlogic/pi-mono (local clone)
 * SHA:     3e5ad67e0f325d4888f82f9b82966218eb4407f5
 * Date:    2026-05-07
 * License: MIT (Copyright (c) 2025 Mario Zechner) — see pi-mono/LICENSE
 * Notes:   Imports rewritten @entelligentsia/* → @entelligentsia/*.
 *          Upstream promotion: see forge-cli/architectural-review.md §R2.
 *
 * Forge-cli divergence (FORGE-S20-T11, must be re-applied on every pi-mono sync):
 *   - `discoverAgents` derives the user agents directory via
 *     `getPiAgentAgentsDir()` from `../paths/paths.js` instead of inlining
 *     `path.join(getAgentDir(), "agents")`. The resolver re-exports pi's
 *     getAgentDir so the resulting path is identical — this is a pure
 *     centralization refactor that keeps AC #1 honest (no `getAgentDir()`
 *     call sites in forge-cli outside the resolver).
 *   - See `forge-cli/vendor-pi/DIVERGENCE.md` (sync-pi-upstream picks it up).
 *
 * Forge-cli divergence (FORGE-S25-T18, must be re-applied on every pi-mono sync):
 *   - Local `isDirectory()` helper removed; import from `../lib/shared-fs-utils.js`
 *     instead. Behaviour is identical (returns false on any error). Finding S-3.
 */
/**
 * Agent discovery and configuration
 */
import * as fs from "node:fs";
import * as path from "node:path";
import { parseFrontmatter } from "@earendil-works/pi-coding-agent";
import { isDirectory } from "../lib/shared-fs-utils.js";
import { getPiAgentAgentsDir } from "../paths/paths.js";
function loadAgentsFromDir(dir, source) {
    const agents = [];
    if (!fs.existsSync(dir)) {
        return agents;
    }
    let entries;
    try {
        entries = fs.readdirSync(dir, { withFileTypes: true });
    }
    catch {
        return agents;
    }
    for (const entry of entries) {
        if (!entry.name.endsWith(".md"))
            continue;
        if (!entry.isFile() && !entry.isSymbolicLink())
            continue;
        const filePath = path.join(dir, entry.name);
        let content;
        try {
            content = fs.readFileSync(filePath, "utf-8");
        }
        catch {
            continue;
        }
        const { frontmatter, body } = parseFrontmatter(content);
        if (!frontmatter.name || !frontmatter.description) {
            continue;
        }
        const tools = frontmatter.tools
            ?.split(",")
            .map((t) => t.trim())
            .filter(Boolean);
        agents.push({
            name: frontmatter.name,
            description: frontmatter.description,
            tools: tools && tools.length > 0 ? tools : undefined,
            model: frontmatter.model,
            systemPrompt: body,
            source,
            filePath,
        });
    }
    return agents;
}
function findNearestProjectAgentsDir(cwd) {
    let currentDir = cwd;
    while (true) {
        const candidate = path.join(currentDir, ".pi", "agents");
        if (isDirectory(candidate))
            return candidate;
        const parentDir = path.dirname(currentDir);
        if (parentDir === currentDir)
            return null;
        currentDir = parentDir;
    }
}
export function discoverAgents(cwd, scope) {
    const userDir = getPiAgentAgentsDir();
    const projectAgentsDir = findNearestProjectAgentsDir(cwd);
    const userAgents = scope === "project" ? [] : loadAgentsFromDir(userDir, "user");
    const projectAgents = scope === "user" || !projectAgentsDir ? [] : loadAgentsFromDir(projectAgentsDir, "project");
    const agentMap = new Map();
    if (scope === "both") {
        for (const agent of userAgents)
            agentMap.set(agent.name, agent);
        for (const agent of projectAgents)
            agentMap.set(agent.name, agent);
    }
    else if (scope === "user") {
        for (const agent of userAgents)
            agentMap.set(agent.name, agent);
    }
    else {
        for (const agent of projectAgents)
            agentMap.set(agent.name, agent);
    }
    return { agents: Array.from(agentMap.values()), projectAgentsDir };
}
export function formatAgentList(agents, maxItems) {
    if (agents.length === 0)
        return { text: "none", remaining: 0 };
    const listed = agents.slice(0, maxItems);
    const remaining = agents.length - listed.length;
    return {
        text: listed.map((a) => `${a.name} (${a.source}): ${a.description}`).join("; "),
        remaining,
    };
}
