// @entelligentsia/forge-compress/src/compressor.ts
import { countTokens } from "./tokens.js";
import {
  stripAnsi,
  stripTimestampsAndHashes,
  normalizeWhitespace,
  isBoilerplate,
} from "./strip.js";

export function lightweightCleanup(content: string): string {
  const lines = content.split("\n");
  const total = lines.length;
  const result: string[] = [];
  let blankCount = 0;
  let braceRun: string[] = [];

  const flushBraceRun = () => {
    if (total <= 200 || braceRun.length <= 5) {
      result.push(...braceRun);
    } else {
      result.push(braceRun[0], braceRun[1]);
      result.push(`[${braceRun.length - 2} brace-only lines collapsed]`);
    }
    braceRun = [];
  };

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) {
      flushBraceRun();
      blankCount++;
      if (blankCount <= 1) result.push("");
      continue;
    }
    blankCount = 0;
    if (/^[}\]);]+$/.test(trimmed)) {
      braceRun.push(trimmed);
      continue;
    }
    flushBraceRun();
    result.push(line);
  }
  flushBraceRun();
  return result.join("\n");
}

export function verbatimCompact(text: string): string {
  const lines: string[] = [];
  let blankCount = 0;
  let prevLine: string | null = null;
  let repeatCount = 0;

  const flushRepeats = () => {
    if (repeatCount > 1 && prevLine !== null) {
      const lastIdx = lines.length - 1;
      if (lastIdx >= 0) {
        lines[lastIdx] = `[${repeatCount}x] ${prevLine}`;
      }
    }
    repeatCount = 0;
    prevLine = null;
  };

  for (const line of text.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) {
      blankCount++;
      if (blankCount <= 1) {
        flushRepeats();
        lines.push("");
      }
      continue;
    }
    blankCount = 0;
    if (isBoilerplate(trimmed)) continue;

    const normalized = normalizeWhitespace(trimmed);
    const stripped = stripTimestampsAndHashes(normalized);

    if (prevLine !== null && prevLine === stripped) {
      repeatCount++;
      continue;
    }
    flushRepeats();
    prevLine = stripped;
    repeatCount = 1;
    lines.push(stripped);
  }
  flushRepeats();
  return lines.join("\n");
}

export function aggressiveCompress(
  content: string,
  ext?: string,
): string {
  const result: string[] = [];
  const isPython = ext === "py";
  const isHtml = ext === "html" || ext === "htm" || ext === "xml" || ext === "svg";
  const isSql = ext === "sql";
  const isShell = ext === "sh" || ext === "bash" || ext === "zsh" || ext === "fish";
  let inBlockComment = false;

  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    if (inBlockComment) {
      if (trimmed.includes("*/") || (isHtml && trimmed.includes("-->"))) {
        inBlockComment = false;
      }
      continue;
    }
    if (trimmed.startsWith("/*") || (isHtml && trimmed.startsWith("<!--"))) {
      if (!(trimmed.includes("*/") || trimmed.includes("-->"))) {
        inBlockComment = true;
      }
      continue;
    }
    if (trimmed.startsWith("//") && !trimmed.startsWith("///")) continue;
    if (trimmed.startsWith("*") || trimmed.startsWith("*/")) continue;
    if (isPython && trimmed.startsWith("#")) continue;
    if (isSql && trimmed.startsWith("--")) continue;
    if (isShell && trimmed.startsWith("#") && !trimmed.startsWith("#!")) continue;

    if (/^[}\]);]+$/.test(trimmed)) {
      const last = result[result.length - 1];
      if (last && /^[}\]);]+$/.test(last.trim())) {
        result[result.length - 1] = last + trimmed;
        continue;
      }
    }
    result.push(trimmed);
  }
  return result.join("\n");
}

export function safeguardRatio(original: string, compressed: string): string {
  const origTokens = countTokens(original);
  const compTokens = countTokens(compressed);
  if (origTokens === 0) return compressed;
  if (compTokens > origTokens) return original;
  const ratio = compTokens / origTokens;
  if (ratio < 0.05 && origTokens < 2000) return original;
  return compressed;
}

export { stripAnsi };
