// @entelligentsia/forge-compress/src/entropy.ts
import { countTokens } from "./tokens.js";

export function shannonEntropy(text: string): number {
  if (!text) return 0;
  const freq = new Map<string, number>();
  for (const ch of text) {
    freq.set(ch, (freq.get(ch) ?? 0) + 1);
  }
  const len = text.length;
  let entropy = 0;
  for (const count of freq.values()) {
    const p = count / len;
    if (p > 0) entropy -= p * Math.log2(p);
  }
  return entropy;
}

export function normalizedTokenEntropy(text: string): number {
  if (!text || text.trim().length === 0) return 0;
  const words = text.trim().split(/\s+/);

  if (words.length <= 1) {
    const charEntropy = shannonEntropy(text.trim());
    return Math.min(1, charEntropy / 4.5);
  }

  const freq = new Map<string, number>();
  for (const w of words) {
    freq.set(w, (freq.get(w) ?? 0) + 1);
  }
  const n = words.length;
  const uniqueRatio = freq.size / n;
  let entropy = 0;
  for (const count of freq.values()) {
    const p = count / n;
    if (p > 0) entropy -= p * Math.log2(p);
  }
  const maxEntropy = Math.log2(n);
  const normalized = maxEntropy > 0 ? entropy / maxEntropy : 0;
  return Math.min(1, normalized * 0.7 + uniqueRatio * 0.3);
}

function renderIb(
  lines: string[],
  scores: number[],
  threshold: number,
): string {
  const out: string[] = [];
  let omitRun = 0;
  for (let i = 0; i < lines.length; i++) {
    if (scores[i] >= threshold) {
      if (omitRun > 0) {
        out.push(`// ... ${omitRun} low-info lines omitted`);
        omitRun = 0;
      }
      out.push(lines[i]);
    } else {
      omitRun++;
    }
  }
  if (omitRun > 0) {
    out.push(`// ... ${omitRun} low-info lines omitted`);
  }
  return out.join("\n");
}

export function compressIb(text: string, targetRatio: number): string {
  if (!text) return "";
  const inputTokens = countTokens(text);
  if (inputTokens === 0) return text;

  const ratio = Math.max(0.02, Math.min(1.0, targetRatio));
  const lines = text.split("\n");
  const scores = lines.map((ln) => normalizedTokenEntropy(ln));

  let lo = 0;
  let hi = 1;
  let best = renderIb(lines, scores, 0);
  let bestDiff = Infinity;

  const consider = (thr: number) => {
    const cand = renderIb(lines, scores, thr);
    const r = countTokens(cand) / inputTokens;
    const diff = Math.abs(r - ratio);
    if (diff < bestDiff) {
      bestDiff = diff;
      best = cand;
    }
    return r;
  };

  for (let i = 0; i < 26; i++) {
    const mid = (lo + hi) / 2;
    const r = consider(mid);
    if (r > ratio) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  for (const thr of [0, 1, lo, hi, (lo + hi) / 2]) {
    consider(thr);
  }
  return best;
}
