// @entelligentsia/forge-compress/src/progressive.ts
import { countTokens, truncateToTokenBudget } from "./tokens.js";

const LAMBDA = 1.35;

function expWeights(n: number): number[] {
  return Array.from({ length: n }, (_, i) => Math.exp(LAMBDA * i));
}

function allocateBudget(budgetTokens: number, weights: number[]): number[] {
  const n = weights.length;
  if (n === 0 || budgetTokens === 0) return new Array(n).fill(0);

  const sumW = Math.max(weights.reduce((a, b) => a + b, 0), Number.EPSILON);
  const base = new Array(n).fill(0);
  const frac = new Array(n).fill(0);

  for (let i = 0; i < n; i++) {
    const exact = (budgetTokens * weights[i]) / sumW;
    base[i] = Math.floor(exact);
    frac[i] = exact - base[i];
  }

  const given = base.reduce((a: number, b: number) => a + b, 0);
  const order = Array.from({ length: n }, (_, i) => i).sort(
    (a, b) => frac[b] - frac[a],
  );

  let extra = budgetTokens - given;
  for (const i of order) {
    if (extra <= 0) break;
    base[i]++;
    extra--;
  }
  return base;
}

function tierForIndex(i: number, n: number): number {
  if (n <= 1) return 2;
  const r = i / (n - 1);
  if (r < 1 / 3) return 0;
  if (r < 2 / 3) return 1;
  return 2;
}

const STRUCTURE_KEYWORDS = [
  "fn ",
  "pub ",
  "struct ",
  "enum ",
  "trait ",
  "impl ",
  "mod ",
  "use ",
  "def ",
  "class ",
  "function ",
  "export ",
  "import ",
  "const ",
  "interface ",
];

function mapLike(s: string, maxTokens: number): string {
  const picked: string[] = [];
  const lines = s.split("\n");
  for (let i = 0; i < lines.length; i++) {
    if (i === 0 || STRUCTURE_KEYWORDS.some((k) => lines[i].includes(k))) {
      picked.push(lines[i]);
    }
    if (picked.length >= 48) break;
  }
  if (picked.length === 0 && lines.length > 0) {
    picked.push(lines[0]);
  }
  return truncateToTokenBudget(picked.join("\n"), Math.max(4, maxTokens));
}

function oneLineSummary(
  segIdx: number,
  s: string,
  maxTokens: number,
): string {
  const preview = s.split("\n")[0]?.slice(0, 120) ?? "";
  const lineCount = s.split("\n").length;
  const draft = `// seg[${segIdx}] ${lineCount} lines, ${s.length} chars | ${preview}`;
  return truncateToTokenBudget(draft, Math.max(8, maxTokens));
}

export function compressProgressive(
  segments: string[],
  budgetTokens: number,
): string[] {
  const n = segments.length;
  if (n === 0) return [];
  if (budgetTokens === 0) return segments.map(() => "");

  const weights = expWeights(n);
  const allocs = allocateBudget(budgetTokens, weights);
  const out: string[] = [];

  for (let i = 0; i < n; i++) {
    const alloc = allocs[i];
    const tier = tierForIndex(i, n);
    if (alloc === 0) {
      out.push("");
      continue;
    }
    let compressed: string;
    switch (tier) {
      case 2:
        compressed = truncateToTokenBudget(segments[i], alloc);
        break;
      case 1:
        compressed = mapLike(segments[i], alloc);
        break;
      default:
        compressed = oneLineSummary(i, segments[i], alloc);
        break;
    }
    out.push(truncateToTokenBudget(compressed, alloc));
  }
  return out;
}
