// @entelligentsia/forge-compress/dist/forge/entity.js
import { countTokens } from "../tokens.js";
const PHASE_ORDER = [
    "plan",
    "triage",
    "review_plan",
    "implementation",
    "code_review",
    "validation",
    "approve",
];
export function compressEntity(input, opts) {
    let entity;
    try {
        entity = JSON.parse(input);
    }
    catch {
        return input;
    }
    const keepSummaries = opts?.keepSummaries ?? "latest";
    const maxKeyChanges = opts?.maxKeyChanges ?? 3;
    const result = {};
    const idKey = findIdKey(entity);
    if (idKey)
        result[idKey] = entity[idKey];
    if (entity.title)
        result.title = entity.title;
    if (entity.status)
        result.status = entity.status;
    if (entity.sprintId)
        result.sprintId = entity.sprintId;
    if (entity.feature_id)
        result.feature_id = entity.feature_id;
    if (entity.severity)
        result.severity = entity.severity;
    if (entity.estimate)
        result.estimate = entity.estimate;
    if (entity.executionMode)
        result.executionMode = entity.executionMode;
    if (entity.pipeline)
        result.pipeline = entity.pipeline;
    if (Array.isArray(entity.taskIds)) {
        result.taskIds = entity.taskIds;
        result.taskCount = entity.taskIds.length;
    }
    if (Array.isArray(entity.dependencies) && entity.dependencies.length > 0) {
        result.dependencies = entity.dependencies;
    }
    const summaries = entity.summaries;
    if (summaries && typeof summaries === "object") {
        if (keepSummaries === "all") {
            result.summaries = compressSummaries(summaries, maxKeyChanges);
        }
        else if (keepSummaries === "latest") {
            const latest = findLatestPhase(summaries);
            if (latest) {
                result.latestPhase = latest.phase;
                result.latestSummary = compressOneSummary(latest.summary, maxKeyChanges);
            }
        }
    }
    if (opts?.flatFormat) {
        return formatFlat(result);
    }
    let output = JSON.stringify(result, null, 2);
    if (opts?.maxTokens && countTokens(output) > opts.maxTokens) {
        output = JSON.stringify(result);
    }
    return output;
}
export function compressEntityList(input, opts) {
    let entities;
    try {
        entities = JSON.parse(input);
    }
    catch {
        return input;
    }
    if (!Array.isArray(entities))
        return input;
    const compressed = entities.map((e) => JSON.parse(compressEntity(JSON.stringify(e), { ...opts, flatFormat: false })));
    let output = JSON.stringify(compressed, null, 2);
    if (opts?.maxTokens && countTokens(output) > opts.maxTokens) {
        output = JSON.stringify(compressed);
    }
    return output;
}
function findIdKey(entity) {
    for (const key of ["taskId", "sprintId", "bugId", "featureId", "id"]) {
        if (entity[key])
            return key;
    }
    return undefined;
}
function compressSummaries(summaries, maxKeyChanges) {
    const result = {};
    for (const phase of PHASE_ORDER) {
        if (summaries[phase]) {
            result[phase] = compressOneSummary(summaries[phase], maxKeyChanges);
        }
    }
    return result;
}
function compressOneSummary(summary, maxKeyChanges) {
    const result = {};
    if (summary.objective)
        result.objective = summary.objective;
    if (summary.verdict)
        result.verdict = summary.verdict;
    if (summary.key_changes && summary.key_changes.length > 0) {
        const shown = summary.key_changes.slice(0, maxKeyChanges);
        const extra = summary.key_changes.length - maxKeyChanges;
        result.key_changes = extra > 0 ? [...shown, `+${extra} more`] : shown;
    }
    if (summary.route)
        result.route = summary.route;
    return result;
}
function findLatestPhase(summaries) {
    for (let i = PHASE_ORDER.length - 1; i >= 0; i--) {
        const phase = PHASE_ORDER[i];
        if (summaries[phase]) {
            return { phase, summary: summaries[phase] };
        }
    }
    return null;
}
function formatFlat(obj, prefix = "") {
    const lines = [];
    for (const [key, value] of Object.entries(obj)) {
        const fullKey = prefix ? `${prefix}.${key}` : key;
        if (value === null || value === undefined)
            continue;
        if (Array.isArray(value)) {
            if (value.length <= 5 && value.every((v) => typeof v === "string")) {
                lines.push(`${fullKey}: ${value.join(", ")}`);
            }
            else {
                lines.push(`${fullKey}: [${value.length} items]`);
            }
        }
        else if (typeof value === "object") {
            lines.push(formatFlat(value, fullKey));
        }
        else {
            lines.push(`${fullKey}: ${value}`);
        }
    }
    return lines.filter(Boolean).join("\n");
}
