// @entelligentsia/forgecli/dist/extensions/forgecli/lib/catalog-loader.js
// lib/catalog-loader.ts — FORGE-S25-T27
//
// Runtime loader for the enum catalog and FSM transition tables bundled in the
// forge-cli payload (dist/forge-payload/.schemas/enum-catalog.json and
// dist/forge-payload/.schemas/transitions/{task,sprint,bug}.json).
//
// Generated by forge/tools/build-enum-catalog.cjs (FORGE-S25-T26) and bundled
// by scripts/build-payload.cjs step 2e1b (FORGE-S25-T27).
//
// Design constraints (IL1, forge-cli-engineer Iron Laws):
//   - No import from forge-init.ts or any top-level handler file (lib → handler
//     is a prohibited dep direction; forge-init.ts imports lib/exec-helpers.ts).
//   - Payload root resolved via import.meta.url (same pattern as forge-init.ts
//     _PKG_ROOT). Tests may supply an explicit bundleRoot to the parametric API.
//   - Fail-loud at module init if the catalog is missing in production (a
//     forge-cli distribution without the catalog is broken by definition).
//
// Iron Laws: IL1 (code only in forge-cli/), IL2 (TypeScript), IL6 (no shell).
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
// forge-cli package root — matches forge-init.ts::_PKG_ROOT calculation.
// Compiled file lives at dist/extensions/forgecli/lib/catalog-loader.js.
// _DIR  = dist/extensions/forgecli/lib/
// ..    = dist/extensions/forgecli/
// ../.. = dist/extensions/
// ../../.. = dist/
// ../../../.. = <pkg-root>/
const _DIR = path.dirname(fileURLToPath(import.meta.url));
const _DIST_DIR = path.resolve(_DIR, "..", "..", "..");
const _PKG_ROOT = path.resolve(_DIST_DIR, "..");
/**
 * Resolve the bundled payload root from the package root.
 * Exported for test access (tests may override with a fixture path instead).
 */
export function resolveBundledPayloadRoot() {
    return path.join(_PKG_ROOT, "dist", "forge-payload");
}
// ── Parametric loaders (testable — caller supplies bundleRoot) ────────────────
/**
 * Load enum-catalog.json from the supplied bundleRoot.
 * Throws with a descriptive message if the file is missing or malformed.
 *
 * @param bundleRoot - absolute path to dist/forge-payload/
 */
export function loadEnumCatalog(bundleRoot) {
    const catalogPath = path.join(bundleRoot, ".schemas", "enum-catalog.json");
    if (!fs.existsSync(catalogPath)) {
        throw new Error(`[catalog-loader] enum-catalog.json not found at ${catalogPath}. ` +
            "Run 'npm run build' to regenerate the payload bundle.");
    }
    try {
        return JSON.parse(fs.readFileSync(catalogPath, "utf8"));
    }
    catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        throw new Error(`[catalog-loader] failed to parse enum-catalog.json: ${msg}`);
    }
}
/**
 * Convert a transition-record array to a Set-based lookup map.
 * @internal
 */
function recordsToSetMap(records) {
    const map = {};
    for (const rec of records) {
        map[rec.from] = new Set(rec.to);
    }
    return map;
}
/**
 * Load a single entity's transition table from the bundled payload.
 * @internal
 */
function loadEntityTransitions(bundleRoot, entity) {
    const filePath = path.join(bundleRoot, ".schemas", "transitions", `${entity}.json`);
    if (!fs.existsSync(filePath)) {
        throw new Error(`[catalog-loader] transitions/${entity}.json not found at ${filePath}. ` +
            "Run 'npm run build' to regenerate the payload bundle.");
    }
    try {
        const records = JSON.parse(fs.readFileSync(filePath, "utf8"));
        return recordsToSetMap(records);
    }
    catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        throw new Error(`[catalog-loader] failed to parse transitions/${entity}.json: ${msg}`);
    }
}
/**
 * Load all three entity transition tables from the bundled payload.
 * Throws if any file is missing or malformed.
 *
 * @param bundleRoot - absolute path to dist/forge-payload/
 */
export function loadTransitions(bundleRoot) {
    return {
        task: loadEntityTransitions(bundleRoot, "task"),
        sprint: loadEntityTransitions(bundleRoot, "sprint"),
        bug: loadEntityTransitions(bundleRoot, "bug"),
    };
}
// ── Module-level constants (lazy-loaded on first access) ─────────────────────
//
// Initialized on first read (not at module import time). This defers the IO to
// the first actual use of the transition tables, which avoids failures during
// vitest module loading in test environments where the build artifact path may
// not be present or may resolve differently before tests run.
//
// In production, these are read on the first hook-dispatcher invocation or
// first transition-guard check — acceptable latency for a one-time cold-read.
//
// Fail-loud: a missing catalog throws a descriptive error at first use (not silently
// wrong). A forge-cli distribution without the catalog is broken by definition.
let _cachedTransitions = null;
function getCachedTransitions() {
    if (_cachedTransitions === null) {
        _cachedTransitions = loadTransitions(resolveBundledPayloadRoot());
    }
    return _cachedTransitions;
}
/** Task status FSM as Set-based lookup (T25 ADR-canonical). Source: enum-catalog. Lazy-loaded. */
export const TASK_TRANSITIONS = new Proxy({}, {
    get(_target, prop) {
        return getCachedTransitions().task[prop];
    },
    has(_target, prop) {
        return prop in getCachedTransitions().task;
    },
    ownKeys(_target) {
        return Object.keys(getCachedTransitions().task);
    },
    getOwnPropertyDescriptor(_target, prop) {
        const val = getCachedTransitions().task[prop];
        if (val === undefined)
            return undefined;
        return { value: val, writable: false, enumerable: true, configurable: true };
    },
});
/** Sprint status FSM as Set-based lookup (T25 ADR-canonical). Source: enum-catalog. Lazy-loaded. */
export const SPRINT_TRANSITIONS = new Proxy({}, {
    get(_target, prop) {
        return getCachedTransitions().sprint[prop];
    },
    has(_target, prop) {
        return prop in getCachedTransitions().sprint;
    },
    ownKeys(_target) {
        return Object.keys(getCachedTransitions().sprint);
    },
    getOwnPropertyDescriptor(_target, prop) {
        const val = getCachedTransitions().sprint[prop];
        if (val === undefined)
            return undefined;
        return { value: val, writable: false, enumerable: true, configurable: true };
    },
});
/** Bug status FSM as Set-based lookup (T25 ADR-canonical). Source: enum-catalog. Lazy-loaded. */
export const BUG_TRANSITIONS = new Proxy({}, {
    get(_target, prop) {
        return getCachedTransitions().bug[prop];
    },
    has(_target, prop) {
        return prop in getCachedTransitions().bug;
    },
    ownKeys(_target) {
        return Object.keys(getCachedTransitions().bug);
    },
    getOwnPropertyDescriptor(_target, prop) {
        const val = getCachedTransitions().bug[prop];
        if (val === undefined)
            return undefined;
        return { value: val, writable: false, enumerable: true, configurable: true };
    },
});
