// @entelligentsia/forgecli/dist/extensions/forgecli/collate.js
// forge:collate — native kickoff handler (FORGE-S21-T10).
//
// Replaces the auto-generated stub previously installed by
// registerAllForgeCommands (forge-commands.ts). Kickoff Shim archetype
// (Pack-04 + Pack-06): single LLM handoff in current context, no fork.
//
// The collator_agent.md workflow has no audience restriction (defaults to
// "any"), so standalone human invocations are allowed — useful for retro
// runs and manual collation triggers.
//
// Iron Laws:
//   IL1 — code only under forge-cli/src/extensions/forgecli/.
//   IL4 — no JSON.stringify-into-subagent dispatch.
//   IL6 — no shell-string interpolation; no spawn calls here.
//   IL7 — every failure path emits ctx.ui.notify and returns; no silent
//         continuation.
import * as fs from "node:fs";
import * as path from "node:path";
import { assertAudience } from "./audience-gate.js";
import { sendKickoff } from "./kickoff.js";
// FORGE-S25-T16: extracted to lib modules (H-1, H-2). Re-exported here for
// backward compatibility with existing test and consumer imports.
import { extractPersonaNames } from "./lib/frontmatter-parser.js";
import { loadPersona, PersonaSkillLoaderError } from "./parsers/persona-skill-loader.js";
import { loadWorkflow, WorkflowLoaderError } from "./parsers/workflow-loader.js";
export { extractPersonaNames };
import { checkMaterialization } from "./lib/manifest-checker.js";
export { checkMaterialization };
export function parseCollateArgs(rawArgs, cwd) {
    const trimmed = (rawArgs ?? "").trim();
    if (trimmed === "") {
        return { mode: "empty", taskRef: "", sourceLabel: "(no input — collator infers sprint from store/context)" };
    }
    if (trimmed.startsWith("@")) {
        const ref = trimmed.slice(1).trim();
        const filePath = path.isAbsolute(ref) ? ref : path.resolve(cwd, ref);
        const seed = fs.readFileSync(filePath, "utf8");
        return { mode: "file", taskRef: seed, sourceLabel: `(seed from file: ${ref})` };
    }
    return { mode: "text", taskRef: trimmed, sourceLabel: "(seed from inline text)" };
}
export function composeKickoff(opts) {
    const { workflowMd, personaIdentity, parsed } = opts;
    const sections = ["# /forge:collate", ""];
    if (personaIdentity.trim().length > 0) {
        sections.push(personaIdentity.trim(), "");
    }
    sections.push("## Dispatch", "", "Run the workflow below. Specifically:", "", "1. Identify the active sprint via `forge_store_query` — do NOT raw-read `.forge/store/`.", "2. Follow the workflow Algorithm verbatim: collate store data into MASTER_INDEX.md, sprint INDEX.md, task INDEX.md files.", "3. Write all collated markdown files using the `write` tool.", "4. Honour Pack-06 Read/Write/Ask/Store discipline: store reads go via `forge_store_query`; in-conversation clarifications use `forge_ask_user`.");
    sections.push("", "---", "", "## Workflow", "", workflowMd.trim(), "", "---");
    if (parsed.mode === "empty") {
        sections.push("", "## Input", "", "(no seed — infer sprint from store context)");
    }
    else {
        sections.push("", `## Input — ${parsed.sourceLabel}`, "", parsed.taskRef.trim());
    }
    return sections.join("\n");
}
// Registration -------------------------------------------------------------
const WORKFLOW_REL_PATH = path.join(".forge", "workflows", "collator_agent.md");
export function registerCollate(pi, options = {}) {
    pi.registerCommand("forge:collate", {
        description: "Run the collator-agent workflow. " +
            "Usage: /forge:collate [@<file> | <sprint-id> | <free-form text>]. " +
            "Empty args → collator infers the active sprint from store context.",
        async handler(args, ctx) {
            const cwd = options.cwd ?? process.cwd();
            const workflowPath = path.join(cwd, WORKFLOW_REL_PATH);
            let workflowMd;
            let workflowAudience;
            try {
                const loaded = loadWorkflow(workflowPath);
                workflowMd = loaded.rawMarkdown;
                workflowAudience = loaded.audience;
            }
            catch (err) {
                if (err instanceof WorkflowLoaderError) {
                    if (err.code === "missing_file") {
                        ctx.ui.notify(`× forge:collate — workflow not found at ${WORKFLOW_REL_PATH}; run /forge:init or /forge:rebuild first.`, "error");
                    }
                    else {
                        ctx.ui.notify(`× forge:collate — workflow load failed (${err.code}): ${err.message}`, "error");
                    }
                    return;
                }
                const e = err;
                ctx.ui.notify(`× forge:collate — failed to read workflow: ${e.message ?? "unknown"}`, "error");
                return;
            }
            let parsed;
            try {
                parsed = parseCollateArgs(args, cwd);
            }
            catch (err) {
                const e = err;
                ctx.ui.notify(`× forge:collate — failed to read seed: ${e.message ?? "unknown"}`, "error");
                return;
            }
            const check = checkMaterialization(workflowPath, workflowMd);
            if (!check.ok) {
                for (const marker of check.missing) {
                    ctx.ui.notify(`× workflow regression: ${marker} not found in ${workflowPath}`, "error");
                }
                return;
            }
            const personas = extractPersonaNames(workflowMd);
            let personaIdentity = "";
            if (personas.length > 0) {
                try {
                    const persona = loadPersona(personas[0], { cwd });
                    personaIdentity = persona.identity;
                }
                catch (err) {
                    if (err instanceof PersonaSkillLoaderError) {
                        ctx.ui.notify(`× forge:collate — persona '${personas[0]}' load failed (${err.code}): ${err.message}`, "error");
                        return;
                    }
                    const e = err;
                    ctx.ui.notify(`× forge:collate — persona load error: ${e.message ?? "unknown"}`, "error");
                    return;
                }
            }
            if (!assertAudience({ workflowName: "collator_agent", audience: workflowAudience }, ctx)) {
                return;
            }
            const kickoff = composeKickoff({
                workflowMd,
                personaIdentity,
                parsed,
            });
            sendKickoff(pi, kickoff);
        },
    });
}
