// @entelligentsia/forgecli/dist/extensions/forgecli/forge-subagent.js
// forge-subagent.ts — in-process Forge subagent harness via pi SDK.
//
// Three channels (per design discussion):
//   spawn     → createAgentSession({ resourceLoader, ... })
//   monitor   → session.subscribe(listener) — assistant messages, tool calls, usage
//   terminate → session.abort() (also via AbortSignal)
//
// "steer" channel intentionally unused: Forge orchestrators dispatch one-shot
// per phase / per task. Revision/iterate = TS state machine spawning a fresh
// session, not mid-stream steering.
//
// Personas read from .forge/personas/*.md (already materialized by /forge:init).
// Optional frontmatter (model, tools, description) — defaults applied when absent.
//
// Iron Laws:
//   IL1 — code only under forge-cli/src/extensions/forgecli/
//   IL7 — every failure path returns a typed result; no silent continuation
import * as fs from "node:fs";
import * as path from "node:path";
import { AuthStorage, createAgentSession, DefaultResourceLoader, getAgentDir, ModelRegistry, parseFrontmatter, SessionManager, } from "@earendil-works/pi-coding-agent";
import { FORGE_TOOL_DISCIPLINE } from "./forge-tools.js";
import { buildProjectOrientation } from "./project-orientation.js";
// ── Persona discovery ─────────────────────────────────────────────────────
/**
 * Load a Forge persona from `.forge/personas/<name>.md`.
 *
 * Frontmatter (optional, all keys may be missing):
 *   description: short role summary
 *   model: pi model id (default: project default)
 *   tools: comma-separated tool list (default: all coding tools)
 *
 * If frontmatter absent, name derives from filename and the entire file body
 * is used as the system prompt.
 */
export function loadForgePersona(name, cwd) {
    const filePath = path.join(cwd, ".forge", "personas", `${name}.md`);
    const raw = fs.readFileSync(filePath, "utf-8");
    const { frontmatter, body } = parseFrontmatter(raw);
    const tools = frontmatter.tools
        ?.split(",")
        .map((t) => t.trim())
        .filter(Boolean);
    return {
        name: frontmatter.name ?? name,
        description: frontmatter.description ?? `Forge ${name} persona`,
        model: frontmatter.model,
        tools: tools && tools.length > 0 ? tools : undefined,
        systemPrompt: body,
        filePath,
    };
}
// ── Empty-usage helper ────────────────────────────────────────────────────
function emptyUsage() {
    return { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, cost: 0, contextTokens: 0, turns: 0 };
}
// ── Run a subagent (spawn + monitor + terminate channels) ────────────────
/**
 * Spawn a Forge subagent in-process via pi SDK and run a task to completion.
 *
 * Returns a SubagentResult after `session.prompt()` resolves. If `signal`
 * fires, the session is aborted and the call returns with exitCode=1 and
 * stopReason="aborted".
 *
 * Usage events are aggregated from `turn_end` (per-turn assistant message
 * usage). Total contextTokens are sourced from the latest turn.
 */
export async function runForgeSubagent(opts) {
    const { persona, task, cwd, signal, onEvent, forgeRoot, cacheSessionId } = opts;
    // Captured at dispatch start so the transcript filename's ISO prefix
    // reflects when the subagent began, not when it finished — preserves
    // natural chronological ordering when a phase runs multiple times.
    const startedAt = new Date();
    // Set FORGE_ROOT in the process environment so the subagent's bash tool
    // can resolve $FORGE_ROOT paths. This is critical for workflow commands
    // that cite $FORGE_ROOT/tools/store-cli.cjs etc. Without it, every
    // subagent that shells out to store-cli fails.
    if (forgeRoot) {
        process.env.FORGE_ROOT = forgeRoot;
    }
    const result = {
        exitCode: 0,
        messages: [],
        usage: emptyUsage(),
        model: persona.model,
    };
    // ── spawn channel ─────────────────────────────────────────────────────
    const cwdAbs = cwd ?? process.cwd();
    // Project orientation is prepended to the persona system prompt for every
    // subagent dispatch — see project-orientation.ts and forge-cli#6.
    const orientation = buildProjectOrientation(cwdAbs);
    const loader = new DefaultResourceLoader({
        cwd: cwdAbs,
        agentDir: getAgentDir(),
        systemPromptOverride: () => {
            return `${orientation}\n${persona.systemPrompt}${FORGE_TOOL_DISCIPLINE}`;
        },
        // By default, inherit the parent's extensions (lean-ctx, ollama-cloud,
        // web-search, etc.) so subagents get the same tool surface. Extensions
        // activate in-process — no heavy subprocess spawning. Context files and
        // prompt templates are suppressed to avoid injecting CLAUDE.md /
        // post-init tour bloat into phase prompts where persona + orientation
        // suffice. When streamFn is set (test-only seam) or noExtensions is
        // explicitly true, suppress discovery to avoid loading globally-installed
        // extensions (latency + environment-dependence in test harnesses).
        noExtensions: opts.noExtensions ?? !!opts.streamFn,
        noSkills: opts.noExtensions ?? !!opts.streamFn,
        noPromptTemplates: true,
        noContextFiles: true,
    });
    await loader.reload();
    const authStorage = AuthStorage.create();
    // Prefer the host session's ModelRegistry — it carries extension-registered
    // providers (ollama-cloud, openrouter, etc.). Without it, the freshly
    // created registry would miss every per-persona model and setModel below
    // would silently fall back to pi's current model. See FORGE-BUG-001 +
    // the modelRegistry doc on RunSubagentOptions.
    const modelRegistry = opts.modelRegistry ?? ModelRegistry.create(authStorage);
    // Refresh in case providers were registered after the session started.
    try {
        modelRegistry.refresh?.();
    }
    catch {
        /* ignore */
    }
    const { session } = await createAgentSession({
        sessionManager: SessionManager.inMemory(),
        authStorage,
        modelRegistry,
        resourceLoader: loader,
        customTools: opts.customTools,
        // Persona frontmatter `tools:` is a name allowlist (read/bash/edit/write/...).
        // Omit field → pi enables default built-ins (read, bash, edit, write).
        tools: persona.tools,
    });
    // Apply prompt-cache session affinity. `createAgentSession` in
    // pi-coding-agent does not yet surface `sessionId` in its options
    // (CreateAgentSessionOptions exposes neither sessionId nor cacheRetention
    // as of @earendil-works/pi-coding-agent@0.74), so we set the underlying
    // Agent's public mutable field directly. The Agent forwards it to the
    // provider per stream call.
    if (cacheSessionId) {
        session.agent.sessionId = cacheSessionId;
    }
    // Test-only seam — see RunSubagentOptions.streamFn doc (forge-cli#17).
    if (opts.streamFn) {
        session.agent.streamFn = opts.streamFn;
    }
    // ── Model routing (Plan 16 Slice 2) ──────────────────────────────────
    // Apply requestedModel via the async setter. Falls through to inherit on
    // any failure — never crashes the phase. result.model/provider are read
    // from the stream after the fact (IL10 invariant preserved).
    if (opts.requestedModel) {
        const m = modelRegistry.find(opts.requestedModel.provider, opts.requestedModel.model);
        if (m) {
            try {
                await session.setModel(m);
            }
            catch (err) {
                const e = err;
                process.stderr.write(`[forge-subagent] setModel failed for ${opts.requestedModel.provider}:${opts.requestedModel.model} (non-fatal, inheriting pi current model): ${e.message ?? "unknown"}\n`);
            }
        }
        else {
            process.stderr.write(`[forge-subagent] requestedModel not in registry: ${opts.requestedModel.provider}:${opts.requestedModel.model} — inheriting pi current model\n`);
        }
    }
    // ── terminate channel ────────────────────────────────────────────────
    const onAbort = () => {
        void session.abort();
    };
    if (signal) {
        if (signal.aborted) {
            onAbort();
        }
        else {
            signal.addEventListener("abort", onAbort, { once: true });
        }
    }
    // ── monitor channel ───────────────────────────────────────────────────
    const unsubscribe = session.subscribe((event) => {
        if (onEvent)
            onEvent(event);
        if (event.type === "turn_end" && event.message) {
            const msg = event.message;
            result.messages.push(msg);
            if (msg.role === "assistant") {
                result.usage.turns++;
                const usage = msg.usage;
                if (usage) {
                    result.usage.input += usage.input ?? 0;
                    result.usage.output += usage.output ?? 0;
                    result.usage.cacheRead += usage.cacheRead ?? 0;
                    result.usage.cacheWrite += usage.cacheWrite ?? 0;
                    result.usage.cost += usage.cost?.total ?? 0;
                    result.usage.contextTokens = usage.totalTokens ?? result.usage.contextTokens;
                }
                if (!result.model && msg.model)
                    result.model = msg.model;
                if (!result.provider && msg.provider)
                    result.provider = msg.provider;
                if (msg.stopReason)
                    result.stopReason = msg.stopReason;
                if (msg.errorMessage)
                    result.errorMessage = msg.errorMessage;
            }
        }
        if (event.type === "turn_end" && event.toolResults) {
            for (const tr of event.toolResults) {
                result.messages.push(tr);
            }
        }
    });
    try {
        await session.prompt(task);
    }
    catch (err) {
        const e = err;
        result.exitCode = 1;
        result.errorMessage = result.errorMessage ?? e.message ?? "session.prompt threw";
    }
    finally {
        unsubscribe();
        if (signal)
            signal.removeEventListener("abort", onAbort);
        session.dispose();
    }
    if (result.stopReason === "error" || result.stopReason === "aborted") {
        result.exitCode = 1;
    }
    // ── Auto-export subagent transcript ─────────────────────────────────
    // Default-on. Filename is ISO-timestamp-prefixed so successive runs
    // of the same phase (e.g. plan → review → plan → review on revision
    // loops) all persist and sort chronologically in directory listings.
    // Location: .forge/transcripts/<entity>/<ISO-compact>__<entity>__<phase>.json
    // .forge/ is .gitignored by the existing .forge/ entry, so transcripts
    // are excluded from version control by default.
    try {
        writeSubagentTranscript({
            cwd: cwdAbs,
            persona: persona.name,
            tag: opts.exportTag,
            result,
            startedAt,
        });
    }
    catch (err) {
        const e = err;
        process.stderr.write(`[forge-subagent] transcript export failed (non-fatal): ${e.message ?? "unknown"}\n`);
    }
    return result;
}
function sanitizeForFilename(s) {
    return s
        .replace(/[^a-zA-Z0-9._-]/g, "_")
        .replace(/\.{2,}/g, "_")
        .slice(0, 80);
}
/**
 * Format a Date as a compact UTC timestamp for transcript filenames.
 * Example: 2026-05-28T13:35:00.123Z → "20260528T133500Z" (no dashes,
 * colons, or fractional seconds). Lexicographic sort matches chronological
 * order, which is the property that makes `ls` show transcripts in run order.
 */
export function formatTranscriptTimestamp(d) {
    const iso = d.toISOString();
    // "2026-05-28T13:35:00.123Z" → "20260528T133500Z"
    return iso.replace(/[-:]/g, "").replace(/\.\d+/, "");
}
/** Derive the subdirectory and filename from the export tag.
 *
 * exportTag format for tasks/bugs/sprints: "<entityId>__<role>"
 *   → dir  = <entityId>  (e.g. "FORGE-S21-T04")
 *   → file = <ISO>__<entityId>__<role>.json
 *
 * Untagged / partial tag: "general" as dir, "<ISO>__<persona>.json" filename.
 */
export function computeTranscriptPath(cwd, persona, tag, startedAt) {
    const base = path.join(cwd, ".forge", "transcripts");
    const ts = formatTranscriptTimestamp(startedAt);
    if (tag) {
        const parts = tag.split("__");
        if (parts.length >= 2) {
            // "<entityId>__<role>" → dir = entityId, file = <ISO>__entityId__role.json
            const entityId = parts[0];
            const role = parts.slice(1).join("__");
            const dir = path.join(base, sanitizeForFilename(entityId));
            const filename = `${ts}__${sanitizeForFilename(entityId)}__${sanitizeForFilename(role)}.json`;
            return path.join(dir, filename);
        }
    }
    // No tag or malformed — put in general/ with persona as filename.
    const dir = path.join(base, "general");
    const filename = `${ts}__${sanitizeForFilename(persona)}.json`;
    return path.join(dir, filename);
}
export function writeSubagentTranscript(opts) {
    const { cwd, persona, tag, result, startedAt } = opts;
    const outPath = computeTranscriptPath(cwd, persona, tag, startedAt);
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    const payload = {
        schema: "forge-subagent-transcript/v1",
        startedAt: startedAt.toISOString(),
        finishedAt: new Date().toISOString(),
        cwd,
        persona,
        tag: tag ?? null,
        model: result.model ?? null,
        provider: result.provider ?? null,
        exitCode: result.exitCode,
        stopReason: result.stopReason ?? null,
        errorMessage: result.errorMessage ?? null,
        usage: result.usage,
        messageCount: result.messages.length,
        messages: result.messages,
    };
    fs.writeFileSync(outPath, JSON.stringify(payload, null, 2), "utf-8");
    return outPath;
}
// ── Final-output helper ──────────────────────────────────────────────────
/** Extract the final assistant text from a SubagentResult. */
export function getFinalOutput(messages) {
    for (let i = messages.length - 1; i >= 0; i--) {
        const msg = messages[i];
        if (msg.role === "assistant") {
            for (const part of msg.content) {
                if (part.type === "text")
                    return part.text;
            }
        }
    }
    return "";
}
