// @entelligentsia/forgecli/dist/extensions/forgecli/init-context.js
// init-context.ts — project-name discovery, project-context.json construction,
// and calibration baseline computation for /forge:init.
//
// Per INIT_PARITY_SPEC.md §9 and PLAN.md Phase C.
// Iron Law 6: no shell-string interpolation.
import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
// ── Project name discovery ─────────────────────────────────────────────────
/**
 * Discover project name from package.json:name, falling back to cwd basename.
 *
 * Tries to read <cwd>/package.json and parse the `name` field. If the file is
 * absent, malformed, or missing the name field, returns `path.basename(cwd)`.
 */
export function discoverProjectName(cwd) {
    const pkgPath = path.join(cwd, "package.json");
    try {
        const raw = fs.readFileSync(pkgPath, "utf8");
        const pkg = JSON.parse(raw);
        if (pkg &&
            typeof pkg === "object" &&
            "name" in pkg &&
            typeof pkg.name === "string") {
            const name = pkg.name;
            if (name.trim())
                return name.trim();
        }
    }
    catch {
        // ENOENT or JSON parse error — fall through to basename
    }
    return path.basename(cwd);
}
/**
 * Build a ProjectContext from discovery results and forge config.
 *
 * Merges discoveryResults with config data. Config project.name/prefix take
 * precedence over discovery if both are present (config was set interactively).
 */
export function buildProjectContext(discoveryResults, config) {
    const name = config.project?.name ?? discoveryResults.projectName ?? "";
    const prefix = config.project?.prefix ?? discoveryResults.prefix ?? "";
    const ctx = {
        project: {
            name,
            prefix,
        },
    };
    if (discoveryResults.description) {
        ctx.project.description = discoveryResults.description;
    }
    if (discoveryResults.stack && discoveryResults.stack.length > 0) {
        ctx.project.stack = discoveryResults.stack;
    }
    const testCmd = discoveryResults.testCommand ?? "";
    const buildCmd = discoveryResults.buildCommand ?? "";
    const lintCmd = discoveryResults.lintCommand ?? "";
    if (testCmd || buildCmd) {
        ctx.project.commands = {
            ...(testCmd ? { test: testCmd } : {}),
            ...(buildCmd ? { build: buildCmd } : {}),
        };
    }
    const hasArch = discoveryResults.backendFramework ??
        discoveryResults.frontendFramework ??
        discoveryResults.database ??
        discoveryResults.deployment ??
        discoveryResults.dataAccess;
    if (hasArch) {
        ctx.architecture = {};
        const frameworks = {};
        if (discoveryResults.backendFramework)
            frameworks.backend = discoveryResults.backendFramework;
        if (discoveryResults.frontendFramework)
            frameworks.frontend = discoveryResults.frontendFramework;
        if (discoveryResults.database)
            frameworks.database = discoveryResults.database;
        if (Object.keys(frameworks).length > 0)
            ctx.architecture.frameworks = frameworks;
        if (discoveryResults.dataAccess)
            ctx.architecture.dataAccess = discoveryResults.dataAccess;
        if (discoveryResults.deployment)
            ctx.architecture.deployment = discoveryResults.deployment;
    }
    const hasWorkflow = testCmd || lintCmd || buildCmd;
    if (hasWorkflow) {
        ctx.workflow = {
            ...(testCmd ? { testCommand: testCmd } : {}),
            ...(lintCmd ? { lintCommand: lintCmd } : {}),
            ...(buildCmd ? { buildCommand: buildCmd } : {}),
        };
    }
    const kbPath = discoveryResults.kbPath ?? config.paths?.engineering ?? "engineering";
    ctx.knowledgeBase = {
        path: kbPath,
        indexFile: `${kbPath}/MASTER_INDEX.md`,
    };
    return ctx;
}
/**
 * Validate a ProjectContext object structurally. Throws with a descriptive
 * error if required fields are missing.
 */
export function validateProjectContext(ctx) {
    if (!ctx || typeof ctx !== "object") {
        throw new Error("project-context validation failed: not an object");
    }
    const record = ctx;
    if (!record.project || typeof record.project !== "object") {
        throw new Error("project-context validation failed: missing required field 'project'");
    }
    const project = record.project;
    if (typeof project.name !== "string" || !project.name.trim()) {
        throw new Error("project-context validation failed: project.name is missing or empty");
    }
    if (typeof project.prefix !== "string" || !project.prefix.trim()) {
        throw new Error("project-context validation failed: project.prefix is missing or empty");
    }
}
/**
 * Write project-context.json to <cwd>/.forge/project-context.json
 */
export function writeProjectContext(cwd, ctx) {
    const filePath = path.join(cwd, ".forge", "project-context.json");
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, JSON.stringify(ctx, null, 2) + "\n", "utf8");
}
/**
 * Compute a calibration baseline snapshot for /forge:init Phase 2.
 *
 * @param cwd - project working directory
 * @param kbPath - relative path to the knowledge base directory (e.g. "engineering")
 * @param bundledPluginVersion - version string from package.json forge.bundledVersion
 */
export function computeCalibrationBaseline(cwd, kbPath, bundledPluginVersion) {
    const baseline = {
        lastCalibrated: new Date().toISOString(),
        version: bundledPluginVersion,
        masterIndexHash: null,
        sprintsCovered: [],
    };
    // Compute SHA-256 of MASTER_INDEX.md if it exists
    const masterIndexPath = path.join(cwd, kbPath, "MASTER_INDEX.md");
    try {
        const content = fs.readFileSync(masterIndexPath, "utf8");
        baseline.masterIndexHash = crypto.createHash("sha256").update(content, "utf8").digest("hex");
    }
    catch {
        // File not yet present — leave as null (init just started Phase 2)
    }
    // Collect completed sprint IDs from .forge/store/sprints/
    // Matches the algorithm used by /forge:calibrate Step 2 auto-init branch.
    const storeSprintsDir = path.join(cwd, ".forge", "store", "sprints");
    try {
        const files = fs.readdirSync(storeSprintsDir).filter((f) => f.endsWith(".json"));
        const completedIds = [];
        for (const f of files) {
            try {
                const raw = fs.readFileSync(path.join(storeSprintsDir, f), "utf8");
                const sprint = JSON.parse(raw);
                if (typeof sprint.sprintId === "string" &&
                    (sprint.status === "done" || sprint.status === "retrospective-done")) {
                    completedIds.push(sprint.sprintId);
                }
            }
            catch {
                // Skip malformed sprint file
            }
        }
        baseline.sprintsCovered = completedIds.sort();
    }
    catch {
        // Store sprints dir not yet created — leave as []
    }
    return baseline;
}
