// @entelligentsia/forgecli/dist/extensions/forgecli/usage-hook.js
// Pi-runtime token telemetry hook — FORGE-S19-T03
//
// Registers a pi.on("message_end") listener that captures per-turn token usage
// from AssistantMessage events. Accumulates usage per phase key and exposes
// flushPhaseUsage() to write sidecars via store-cli record-usage.
//
// Phase key is provided by the caller via the getPhaseKey() option. The sprint
// runner sets FORGE_PHASE_KEY before spawning the pi session; individual
// commands can override with their own getPhaseKey implementation.
//
// Non-blocking contract:
//   - Non-assistant messages (user, toolResult) are silently skipped.
//   - Missing/malformed usage field on an assistant message → treated as zeros.
//   - subprocess failure in flushPhaseUsage → stderr warn, no throw.
//
// Iron Law 6 compliance: spawnSync is called with an argv array — no shell
// string interpolation anywhere in this module.
import { spawnSync } from "node:child_process";
import * as path from "node:path";
// ── Internal helpers ──────────────────────────────────────────────────────────
function defaultGetPhaseKey() {
    return process.env.FORGE_PHASE_KEY ?? "default";
}
function safeNumber(val) {
    if (typeof val === "number" && Number.isFinite(val))
        return val;
    return 0;
}
// ── Public API ────────────────────────────────────────────────────────────────
/**
 * Register a message_end listener on `pi` that accumulates per-turn token usage.
 *
 * @returns The accumulator map (keyed by phase key). Pass this to flushPhaseUsage.
 */
export function registerUsageHook(pi, options = {}) {
    const getPhaseKey = options.getPhaseKey ?? defaultGetPhaseKey;
    const accumulator = new Map();
    pi.on("message_end", (event) => {
        const msg = event.message;
        // Only assistant messages carry token usage.
        if (msg.role !== "assistant")
            return;
        const usage = msg.usage;
        const input = safeNumber(usage?.input);
        const output = safeNumber(usage?.output);
        const cacheRead = safeNumber(usage?.cacheRead);
        const cacheWrite = safeNumber(usage?.cacheWrite);
        const costTotal = safeNumber(usage?.cost?.total);
        const model = typeof msg.model === "string" ? msg.model : "unknown";
        const phaseKey = getPhaseKey();
        const existing = accumulator.get(phaseKey);
        if (existing) {
            existing.inputTokens += input;
            existing.outputTokens += output;
            existing.cacheReadTokens += cacheRead;
            existing.cacheWriteTokens += cacheWrite;
            existing.estimatedCostUSD += costTotal;
            existing.model = model; // use last model seen
            existing.turnCount += 1;
        }
        else {
            accumulator.set(phaseKey, {
                inputTokens: input,
                outputTokens: output,
                cacheReadTokens: cacheRead,
                cacheWriteTokens: cacheWrite,
                estimatedCostUSD: costTotal,
                model,
                turnCount: 1,
            });
        }
    });
    return accumulator;
}
/**
 * Flush accumulated usage for `phaseKey` to store-cli record-usage.
 *
 * Non-blocking: if the subprocess exits non-zero, emits a warn line to
 * stderr and returns normally. Never throws.
 */
export function flushPhaseUsage(opts) {
    const { sprintId, eventId, phaseKey, forgeRoot, accumulator } = opts;
    const acc = accumulator.get(phaseKey);
    if (!acc) {
        process.stderr.write(`[warn] usage hook: no accumulated data for phase key "${phaseKey}" — skipping flush\n`);
        return;
    }
    const storeCli = path.join(forgeRoot, "tools", "store-cli.cjs");
    // Iron Law 6: argv array — no shell-string interpolation.
    const argv = [
        storeCli,
        "record-usage",
        sprintId,
        eventId,
        "--input-tokens",
        String(acc.inputTokens),
        "--output-tokens",
        String(acc.outputTokens),
        "--cache-read-tokens",
        String(acc.cacheReadTokens),
        "--cache-write-tokens",
        String(acc.cacheWriteTokens),
        "--estimated-cost-usd",
        String(acc.estimatedCostUSD),
        "--token-source",
        "reported",
        "--model",
        acc.model,
    ];
    const result = spawnSync(process.execPath, argv, {
        cwd: process.cwd(),
        stdio: "pipe",
    });
    if (result.status !== 0) {
        const errText = result.stderr ? String(result.stderr).trim() : "(no stderr)";
        process.stderr.write(`[warn] usage hook: record-usage exited ${result.status ?? "null"} for phase "${phaseKey}" — ${errText}\n`);
        return;
    }
    if (opts.onFlush) {
        opts.onFlush(phaseKey, acc);
    }
}
