// @entelligentsia/forgecli/dist/extensions/forgecli/viewport-events.js
// viewport-events.ts
//
// Reusable subagent event observer. Wires `runForgeSubagent.onEvent` to:
//   1. Append per-event tail lines (with tree connectors, glyphs, risky tag,
//      thinking one-liners, batch markers) into SessionRegistry.
//   2. Maintain cumulative phase token usage and push it to the registry on
//      every turn_end — drives the sticky footer in the tail view AND the
//      aggregate Σ meter in the chip strip.
//   3. Surface counts (turn / tools / errors) so the orchestrator can render
//      its end-of-phase summary.
//
// Used by run-task, fix-bug, and run-sprint (ceremony) so every subagent
// "bubbles" usage up the same way. Top-level viewports read aggregate usage
// via SessionRegistry.getAggregateUsage().
import { getOrchestratorTree } from "./orchestrator-tree.js";
import { extractThinkingOneLiner, extractTurnPreview, argHint as fmtArgHint, RISKY_TAG, readUsage, resultShape, toolGlyph, } from "./viewport-renderer.js";
function extractCompression(result) {
    if (!result || typeof result !== "object")
        return undefined;
    const r = result;
    const c = r.details?.compression;
    if (!c || typeof c.before !== "number" || typeof c.after !== "number")
        return undefined;
    return { tool: c.tool ?? "", before: c.before, after: c.after, saved: c.saved ?? 0 };
}
export function attachViewportObserver(opts) {
    const { registry, sessionId, phaseRole, displayRole, beginHeader, writeDebug, setStatusVerbose, notify, verboseKeys, afterEach, } = opts;
    const role = displayRole ?? phaseRole;
    const tree = getOrchestratorTree();
    // Helper to resolve the correct active node ID in the OrchestratorTree.
    const resolveNodeId = () => {
        const exactId = `${sessionId}:${phaseRole}`;
        const exactNode = tree.getNode(exactId);
        if (exactNode && exactNode.status === "running")
            return exactId;
        const parentNode = tree.getNode(sessionId);
        if (parentNode) {
            const prefix = `${sessionId}:${phaseRole}:`;
            for (const cid of parentNode.children) {
                if (cid.startsWith(prefix)) {
                    const child = tree.getNode(cid);
                    if (child && child.status === "running") {
                        return cid;
                    }
                }
            }
        }
        return exactId;
    };
    const state = {
        turn: 0,
        toolCount: 0,
        errCount: 0,
        lastTool: "",
        cumUsage: { input: 0, output: 0, cacheRead: 0 },
        cumCompression: { calls: 0, tokensSaved: 0 },
    };
    // Per-turn tree-connector state.
    let toolsThisTurn = 0;
    let firstLineOfTurn = true;
    let currentTurnPrefixWidth = 0;
    const argsByCallId = new Map();
    const formatTime = (ms) => {
        const d = new Date(ms);
        const pad = (n) => String(n).padStart(2, "0");
        return `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
    };
    const tailPrefix = () => `[${role} ${formatTime(Date.now())} t${state.turn}]`;
    const appendTail = (line, opts) => {
        registry.appendTail(sessionId, phaseRole, line, opts);
        tree.appendTail(resolveNodeId(), line, opts);
    };
    const extractErrorSummary = (result) => {
        const raw = typeof result === "string"
            ? result
            : typeof result === "object" && result !== null
                ? JSON.stringify(result)
                : String(result);
        const firstLine = raw.split(/\r?\n/).find((l) => l.trim().length > 0) ?? raw;
        return firstLine.length > 160 ? `${firstLine.slice(0, 160)}…` : firstLine;
    };
    const emitTurnLine = (body, lineOpts) => {
        const isFirst = firstLineOfTurn;
        const isClosing = lineOpts?.closing === true;
        let branch;
        if (isFirst && isClosing)
            branch = "─";
        else if (isFirst)
            branch = "╭";
        else if (isClosing)
            branch = "╰";
        else
            branch = "│";
        let prefix;
        if (isFirst) {
            prefix = tailPrefix();
            currentTurnPrefixWidth = prefix.length;
        }
        else {
            prefix = " ".repeat(currentTurnPrefixWidth);
        }
        firstLineOfTurn = false;
        appendTail(`${prefix} ${branch} ${body}`, lineOpts?.warning ? { warning: true } : undefined);
    };
    if (beginHeader)
        appendTail(`${tailPrefix()} ${beginHeader}`);
    const onEvent = (event) => {
        switch (event.type) {
            case "turn_start": {
                state.turn++;
                state.lastTool = "";
                toolsThisTurn = 0;
                firstLineOfTurn = true;
                registry.bumpTurn(sessionId);
                tree.bumpNodeTurn(resolveNodeId());
                break;
            }
            case "turn_end": {
                const e = event;
                const delta = readUsage(e.message);
                state.cumUsage.input += delta.input;
                state.cumUsage.output += delta.output;
                state.cumUsage.cacheRead += delta.cacheRead;
                registry.setPhaseUsage(sessionId, phaseRole, state.cumUsage);
                const nodeId = resolveNodeId();
                tree.setNodeUsage(nodeId, state.cumUsage);
                // Capture which (provider, model) actually served this turn so
                // the per-phase tail-view footer can display it. Cheap; the
                // registry no-ops when the value hasn't changed.
                const msg = e.message;
                if (msg && (msg.provider || msg.model)) {
                    registry.setPhaseModel(sessionId, phaseRole, {
                        provider: msg.provider,
                        model: msg.model,
                    });
                    tree.setNodeModel(nodeId, msg.model ?? "", msg.provider ?? "");
                }
                const closingBodies = [];
                const thinking = extractThinkingOneLiner(e.message);
                if (thinking)
                    closingBodies.push(`✱ ${thinking}`);
                const preview = extractTurnPreview(e.message);
                if (preview) {
                    registry.setTurnPreview(sessionId, preview);
                    tree.setTurnPreview(nodeId, preview);
                    if (setStatusVerbose && verboseKeys?.messageKey) {
                        setStatusVerbose(verboseKeys.messageKey, `  "${preview}"`);
                    }
                    closingBodies.push(`» "${preview}"`);
                }
                if (toolsThisTurn > 1) {
                    closingBodies.push(`⇉ batched ${toolsThisTurn} tool calls in turn ${state.turn}`);
                }
                for (let i = 0; i < closingBodies.length; i++) {
                    const isLast = i === closingBodies.length - 1;
                    emitTurnLine(closingBodies[i], { closing: isLast });
                }
                // Bubble one consolidated event up to the registry so the
                // top-level viewport can surface activity from any subagent
                // with subagent-identified attribution.
                registry.recordTurnEvent({
                    sessionId,
                    phaseRole,
                    displayRole: role,
                    turn: state.turn,
                    preview: preview ?? "",
                    thinking: thinking ?? "",
                    deltaUsage: { ...delta },
                    cumUsage: { ...state.cumUsage },
                    timestamp: Date.now(),
                });
                break;
            }
            case "tool_execution_start": {
                const e = event;
                state.toolCount++;
                toolsThisTurn++;
                argsByCallId.set(e.toolCallId, e.args);
                const hint = fmtArgHint(e.toolName, e.args);
                const { glyph, risky } = toolGlyph(e.toolName, e.args);
                state.lastTool = `${e.toolName}${hint ? ` ${hint}` : ""}`;
                writeDebug?.({
                    kind: "tool_start",
                    toolName: e.toolName,
                    toolCallId: e.toolCallId,
                    args: e.args,
                });
                registry.recordToolStart(sessionId, e.toolCallId, e.toolName, e.args);
                tree.incrementNodeToolCount(resolveNodeId());
                const riskPrefix = risky ? `${RISKY_TAG} ` : "";
                emitTurnLine(`${riskPrefix}${glyph} ${e.toolName}${hint ? ` ${hint}` : ""}`, risky ? { warning: true } : undefined);
                break;
            }
            case "tool_execution_end": {
                const e = event;
                argsByCallId.delete(e.toolCallId);
                writeDebug?.({
                    kind: "tool_end",
                    toolName: e.toolName,
                    toolCallId: e.toolCallId,
                    isError: e.isError,
                    result: e.result,
                });
                registry.recordToolEnd(sessionId, e.toolCallId, e.toolName, e.isError, e.result);
                const nodeId = resolveNodeId();
                if (e.isError) {
                    state.errCount++;
                    tree.incrementNodeErrCount(nodeId);
                    emitTurnLine(`⚠ ${e.toolName} failed: ${extractErrorSummary(e.result)}`, { warning: true });
                }
                else {
                    const shape = resultShape(e.toolName, e.result);
                    const comp = extractCompression(e.result);
                    const compHint = comp ? ` ⇌${comp.saved}%` : "";
                    emitTurnLine(`← ${e.toolName} ok${shape ? ` ${shape}` : ""}${compHint}`);
                    if (comp) {
                        state.cumCompression.calls++;
                        state.cumCompression.tokensSaved += comp.before - comp.after;
                        registry.setPhaseCompression(sessionId, phaseRole, state.cumCompression);
                        tree.setNodeCompression(nodeId, state.cumCompression);
                    }
                }
                break;
            }
            case "compaction_start": {
                const e = event;
                notify?.(`◌ ${role}: context compacting (${e.reason})…`, "info");
                appendTail(`${tailPrefix()} ◌ compacting (${e.reason})`);
                break;
            }
            case "auto_retry_start": {
                const e = event;
                const err = e.errorMessage ?? "";
                notify?.(`↻ ${role}: model retry ${e.attempt}/${e.maxAttempts}${err ? `\n${err}` : ""}`, "warning");
                appendTail(`${tailPrefix()} ↻ retry ${e.attempt}/${e.maxAttempts}${err ? `: ${extractErrorSummary(err)}` : ""}`, { warning: true });
                break;
            }
        }
        afterEach?.();
    };
    return { onEvent, state };
}
